// react-vant/es/card/Card.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import cls from 'clsx';
import { BORDER_BOTTOM, BORDER_TOP } from '../utils/constant';
import { createNamespace } from '../utils';
export const CardHeader = props => {
  const [bem] = createNamespace('card-header');
  return _jsxs("div", Object.assign({
    className: cls(props.className, bem(), {
      [BORDER_BOTTOM]: props.border
    }),
    style: props.style,
    onClick: props.onClick
  }, {
    children: [_jsx("div", Object.assign({
      className: cls(bem('content'))
    }, {
      children: props.children
    })), props.extra && _jsx("div", Object.assign({
      className: cls(bem('extra'))
    }, {
      children: props.extra
    }))]
  }));
};
export const CardBody = props => {
  const [bem] = createNamespace('card-body');
  return _jsx("div", Object.assign({
    className: cls(props.className, bem()),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};
export const CardFooter = props => {
  const [bem] = createNamespace('card-footer');
  return _jsx("div", Object.assign({
    className: cls(props.className, bem({
      compact: props.compact
    }), {
      [BORDER_TOP]: props.border
    }),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};
export const CardCover = props => {
  const [bem] = createNamespace('card-cover');
  return _jsx("div", Object.assign({
    className: cls(props.className, bem()),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};

const Card = props => {
  const [bem] = createNamespace('card');
  const {
    className,
    style,
    round,
    border,
    children
  } = props;
  return _jsx("div", Object.assign({
    className: cls(bem({
      round,
      border
    }), className),
    style: style,
    onClick: props.onClick
  }, {
    children: children
  }));
}; // defaultProps defined if need


Card.defaultProps = {};
export default Card;
