// react-vant/lib/floating-ball/useFloatingTouch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

var _hooks = require("../hooks");

var _utils = require("../utils");

var _raf = require("../utils/raf");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function getDirection(x, y) {
  if (x > y) {
    return 'horizontal';
  }

  if (y > x) {
    return 'vertical';
  }

  return '';
}

const INITIAL_STATE = {
  startX: 0,
  startY: 0,
  deltaX: 0,
  deltaY: 0,
  offsetX: 0,
  offsetY: 0,
  _pointX: 0,
  _pointY: 0,

  /** 初始化偏移量计算渲染是否完成 */
  ready: false,
  direction: ''
};

const useFloatingTouch = props => {
  const {
    boundary = true
  } = props;
  const [state, update] = (0, _hooks.useSetState)(INITIAL_STATE);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    if (!props.target) return; // Fill the initial offset state

    const rect = props.target.getBoundingClientRect();
    const {
      width,
      height
    } = rect;
    const offset = Object.entries(props.offset).reduce((a, [k, v]) => {
      let px = (0, _utils.unitToPx)(v);
      let axis = 'deltaY';
      if (k === 'left') px = window.innerWidth - px - width;
      if (k === 'top') px = window.innerHeight - px - height;
      if (k === 'left' || k === 'right') axis = 'deltaX'; // 右下角为坐标原点 取反值

      a[axis] = -px;
      return a;
    }, {});
    update(Object.assign({}, offset));
    (0, _raf.raf)(() => {
      update({
        ready: true
      });
    });
  }, [props.target]);
  const getBound = (0, _react().useCallback)(() => {
    if (!props.target) return {};
    return props.target.getBoundingClientRect();
  }, [props.target]);
  const isVertical = (0, _react().useCallback)(() => state.direction === 'vertical', [state.direction]);
  const isHorizontal = (0, _react().useCallback)(() => state.direction === 'horizontal', [state.direction]);

  const reset = () => {
    update(INITIAL_STATE);
  };

  const start = event => {
    const rect = getBound();
    update(v => ({
      startX: event.touches[0].clientX - v.deltaX,
      startY: event.touches[0].clientY - v.deltaY,
      _pointX: event.touches[0].clientX - rect.left,
      _pointY: event.touches[0].clientY - rect.top
    }));
  };

  const move = event => {
    const touch = event.touches[0];
    const rect = getBound();
    update(value => {
      // Fix: Safari back will set clientX to negative number
      const newState = Object.assign({}, value);
      newState.deltaX = touch.clientX - newState.startX;
      if (boundary && touch.clientX - newState._pointX <= 0) newState.deltaX = -window.innerWidth + rect.width;
      if (boundary && touch.clientX - newState._pointX >= window.innerWidth - rect.width) newState.deltaX = 0;
      newState.deltaY = touch.clientY - newState.startY;
      if (boundary && touch.clientY - newState._pointY <= 0) newState.deltaY = -window.innerHeight + rect.height;
      if (boundary && touch.clientY - newState._pointY >= window.innerHeight - rect.height) newState.deltaY = 0;
      newState.offsetX = Math.abs(newState.deltaX);
      newState.offsetY = Math.abs(newState.deltaY); // lock direction when distance is greater than a certain value

      const LOCK_DIRECTION_DISTANCE = 10;

      if (!newState.direction || newState.offsetX < LOCK_DIRECTION_DISTANCE && newState.offsetY < LOCK_DIRECTION_DISTANCE) {
        newState.direction = getDirection(newState.offsetX, newState.offsetY);
      }

      return newState;
    });
  };

  return Object.assign(Object.assign({}, state), {
    update,
    move,
    start,
    reset,
    isVertical,
    isHorizontal
  });
};

var _default = useFloatingTouch;
exports.default = _default;
