// react-vant/lib/utils/base.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extend = void 0;
exports.get = get;
exports.inBrowser = void 0;
exports.isDef = isDef;
exports.isFunction = isFunction;
exports.isObject = isObject;
exports.isPromise = isPromise;
exports.noop = noop;
exports.once = once;
exports.pick = pick;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable @typescript-eslint/no-explicit-any */
function noop() {}

const extend = Object.assign;
exports.extend = extend;
const inBrowser = typeof window !== 'undefined';
exports.inBrowser = inBrowser;

function isDef(val) {
  return val !== undefined && val !== null;
} // eslint-disable-next-line @typescript-eslint/ban-types


function isFunction(val) {
  return typeof val === 'function';
}

function isPromise(val) {
  return isObject(val) && isFunction(val.then) && isFunction(val.catch);
}

function isObject(val) {
  return val !== null && typeof val === 'object';
}

function get(object, path) {
  const keys = path.split('.');
  let result = object;
  keys.forEach(key => {
    var _a;

    result = (_a = result[key]) !== null && _a !== void 0 ? _a : '';
  });
  return result;
}

function pick(obj, keys, ignoreUndefined) {
  return keys.reduce((ret, key) => {
    if (!ignoreUndefined || obj[key] !== undefined) {
      ret[key] = obj[key];
    }

    return ret;
  }, {});
}

function once(fn) {
  return (...args) => {
    if (!fn) return;
    fn(...args);
    fn = null;
  };
}
