// react-vant/lib/cascader/Cascader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

function _icons() {
  const data = require("@react-vant/icons");

  _icons = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

var _hooks = require("../hooks");

var _tabs = _interopRequireDefault(require("../tabs"));

var _ConfigProviderContext = _interopRequireDefault(require("../config-provider/ConfigProviderContext"));

var _popup = _interopRequireDefault(require("../popup"));

var _useCascaderExtend = require("./useCascaderExtend");

var _useDebunceEffect = _interopRequireDefault(require("../hooks/use-debunce-effect"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint-disable no-plusplus */
const [bem] = (0, _utils.createNamespace)('cascader');

const Cascader = props => {
  const {
    locale
  } = (0, _react().useContext)(_ConfigProviderContext.default);
  const [value, setValue] = (0, _react().useState)(() => props.value === undefined ? props.defaultValue : props.value);
  const [activeTab, updateActiveTab] = (0, _react().useState)(0);
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = (0, _utils.extend)({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.fieldNames);
  const {
    tabs,
    items,
    depth
  } = (0, _useCascaderExtend.useCascaderExtend)(props.options, {
    textKey,
    valueKey,
    childrenKey
  }, value || []); // sync props.value to inner value

  (0, _hooks.useUpdateEffect)(() => {
    if (props.value === undefined) return; // uncontroll mode

    if (JSON.stringify(value) !== JSON.stringify(props.value)) {
      setValue(props.value);
    }
  }, [props.value]); // update active tab index from value

  (0, _react().useEffect)(() => {
    let tabIndex = 0;
    if (Array.isArray(value) && value.length > 0) tabIndex = value.length;
    if (tabIndex >= depth) tabIndex = depth - 1;
    if (tabIndex === activeTab) return;
    updateActiveTab(tabIndex);
  }, [value]);
  (0, _useDebunceEffect.default)(() => {
    var _a, _b;

    if (JSON.stringify(props.value) === JSON.stringify(value)) return;
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, value, items);

    if (value.length >= depth) {
      (_b = props.onFinish) === null || _b === void 0 ? void 0 : _b.call(props, value, items);
    }
  }, [value], {
    wait: 0,
    leading: false,
    trailing: true
  });

  const onSelect = (option, tabIndex) => {
    if (option.disabled) {
      return;
    }

    setValue(prev => {
      const next = [...prev];
      next[tabIndex] = option === null || option === void 0 ? void 0 : option[valueKey];
      return next.slice(0, tabIndex + 1);
    });
  };

  const onClickTab = ({
    name,
    index
  }) => {
    var _a;

    (_a = props.onClickTab) === null || _a === void 0 ? void 0 : _a.call(props, +name);
    updateActiveTab(index);
  };

  const renderCloseIcon = () => {
    if (!props.closeable) return null;

    if (props.closeIcon) {
      return _react().default.cloneElement(props.closeIcon, {
        className: (0, _clsx().default)(bem('close-icon')),
        onClick: props.onClose
      });
    }

    return (0, _jsxRuntime().jsx)(_icons().Cross, {
      className: (0, _clsx().default)(bem('close-icon')),
      onClick: props.onClose
    });
  };

  const renderHeader = () => (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(bem('header'))
  }, {
    children: [(0, _jsxRuntime().jsx)("h2", Object.assign({
      className: (0, _clsx().default)(bem('title'))
    }, {
      children: props.title
    })), renderCloseIcon()]
  }));

  const renderOption = (option, selected, tabIndex) => {
    const color = option.color || (selected ? props.activeColor : undefined);
    const Text = props.optionRender ? props.optionRender({
      option,
      selected
    }) : (0, _jsxRuntime().jsx)("span", {
      children: option[textKey]
    });
    return (0, _jsxRuntime().jsxs)("li", Object.assign({
      className: (0, _clsx().default)(bem('option', {
        selected,
        disabled: option.disabled
      }), option.className),
      style: {
        color
      },
      onClick: () => onSelect(option, tabIndex)
    }, {
      children: [Text, selected ? (0, _jsxRuntime().jsx)(_icons().Success, {
        className: (0, _clsx().default)(bem('selected-icon'))
      }) : null]
    }), option[valueKey]);
  };

  const renderOptions = (options, selectedOption, tabIndex) => (0, _jsxRuntime().jsx)("ul", Object.assign({
    className: (0, _clsx().default)(bem('options'))
  }, {
    children: options.map(option => renderOption(option, option[valueKey] === (selectedOption === null || selectedOption === void 0 ? void 0 : selectedOption[valueKey]), tabIndex))
  }), tabIndex);

  const renderTab = (options, tabIndex) => {
    const selectedOption = items[tabIndex];
    const title = selectedOption ? selectedOption[textKey] : props.placeholder || locale.vanPicker.select;
    return (0, _jsxRuntime().jsx)(_tabs.default.TabPane, Object.assign({
      title: title,
      titleClass: (0, _clsx().default)(bem('tab', {
        unselected: !selectedOption
      }))
    }, {
      children: renderOptions(options, selectedOption, tabIndex)
    }), tabIndex);
  };

  const renderTabs = () => (0, _jsxRuntime().jsx)(_tabs.default, Object.assign({
    animated: true,
    active: activeTab,
    className: (0, _clsx().default)(bem('tabs')),
    color: props.activeColor,
    swipeThreshold: 0,
    swipeable: props.swipeable,
    duration: 300,
    onClickTab: onClickTab
  }, {
    children: tabs.map(renderTab)
  }));

  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(bem())
  }, {
    children: [renderHeader(), tabs.length ? renderTabs() : null]
  }));
};

const CascaderPopup = _react().default.forwardRef((props, ref) => {
  var _a;

  const {
    visible: outerVisible,
    popup
  } = props,
        cascaderProps = (0, _tslib().__rest)(props, ["visible", "popup"]);
  const [visible, setVisible] = (0, _hooks.usePropsValue)({
    value: outerVisible,
    defaultValue: false
  });
  const [value, setValue] = (0, _react().useState)(() => props.value === undefined ? props.defaultValue : props.value);
  const [innerValue, setInnerValue] = (0, _react().useState)(value);
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = (0, _utils.extend)({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.fieldNames);
  const {
    items
  } = (0, _useCascaderExtend.useCascaderExtend)(props.options, {
    textKey,
    valueKey,
    childrenKey
  }, value || []); // sync props.value to inner value

  (0, _hooks.useUpdateEffect)(() => {
    if (props.value === undefined) return; // uncontroll mode

    if (JSON.stringify(value) !== JSON.stringify(props.value)) {
      setValue(props.value);
    }
  }, [props.value]); // sync value to cascader value

  (0, _react().useEffect)(() => {
    if (!popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [value]); // only sync value to popup cascader value when visible change

  (0, _react().useEffect)(() => {
    if (popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [visible]);
  const actions = {
    toggle: () => {
      if (popup) setVisible(v => !v);
    },
    open: () => {
      if (popup) {
        setVisible(true);
      }
    },
    close: () => {
      if (popup) {
        setVisible(false);
      }
    }
  };

  const onClose = () => {
    var _a;

    (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
    actions.close();
  };

  const onFinish = (0, _hooks.useMemoizedFn)((val, selectedRows) => {
    var _a;

    setValue(val);
    (_a = props.onFinish) === null || _a === void 0 ? void 0 : _a.call(props, val, selectedRows);
    actions.close();
  });
  const onChange = (0, _hooks.useMemoizedFn)((val, selectedRows) => {
    var _a, _b;

    setInnerValue(val);

    if (popup) {
      if (visible) (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, val, selectedRows);
    } else {
      (_b = props.onChange) === null || _b === void 0 ? void 0 : _b.call(props, val, selectedRows);
    }
  });
  (0, _react().useImperativeHandle)(ref, () => actions);
  const content = (0, _jsxRuntime().jsx)(Cascader, Object.assign({
    value: innerValue
  }, cascaderProps, {
    onChange: onChange,
    onFinish: onFinish,
    onClose: onClose
  }));
  if (!popup) return content;
  const popupProps = (0, _utils.isObject)(popup) ? Object.assign({
    closeOnClickOverlay: true
  }, popup) : {
    closeOnClickOverlay: true
  };
  return (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
    children: [(0, _jsxRuntime().jsx)(_popup.default, Object.assign({
      position: 'bottom',
      visible: visible,
      closeOnClickOverlay: true,
      onClickOverlay: () => {
        if (!(popupProps === null || popupProps === void 0 ? void 0 : popupProps.closeOnClickOverlay)) return;
        setVisible(false);
      }
    }, popupProps, {
      children: content
    })), (_a = props.children) === null || _a === void 0 ? void 0 : _a.call(props, value, items, actions)]
  });
});

CascaderPopup.defaultProps = {
  closeable: true,
  swipeable: false,
  defaultValue: [],
  options: []
};
var _default = CascaderPopup;
exports.default = _default;
