// react-vant/lib/card/Card.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CardHeader = exports.CardFooter = exports.CardCover = exports.CardBody = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _constant = require("../utils/constant");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const CardHeader = props => {
  const [bem] = (0, _utils.createNamespace)('card-header');
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem(), {
      [_constant.BORDER_BOTTOM]: props.border
    }),
    style: props.style,
    onClick: props.onClick
  }, {
    children: [(0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('content'))
    }, {
      children: props.children
    })), props.extra && (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('extra'))
    }, {
      children: props.extra
    }))]
  }));
};

exports.CardHeader = CardHeader;

const CardBody = props => {
  const [bem] = (0, _utils.createNamespace)('card-body');
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem()),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};

exports.CardBody = CardBody;

const CardFooter = props => {
  const [bem] = (0, _utils.createNamespace)('card-footer');
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem({
      compact: props.compact
    }), {
      [_constant.BORDER_TOP]: props.border
    }),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};

exports.CardFooter = CardFooter;

const CardCover = props => {
  const [bem] = (0, _utils.createNamespace)('card-cover');
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem()),
    style: props.style,
    onClick: props.onClick
  }, {
    children: props.children
  }));
};

exports.CardCover = CardCover;

const Card = props => {
  const [bem] = (0, _utils.createNamespace)('card');
  const {
    className,
    style,
    round,
    border,
    children
  } = props;
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(bem({
      round,
      border
    }), className),
    style: style,
    onClick: props.onClick
  }, {
    children: children
  }));
}; // defaultProps defined if need


Card.defaultProps = {};
var _default = Card;
exports.default = _default;
