// react-vant/es/typography/Ellipsis.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import clsx from 'clsx';
import { useMemo, useRef, useState } from 'react';
import { useIsomorphicLayoutEffect } from '../hooks';
import { useResizeEffect } from '../hooks/use-resize-effect';
import { withStopPropagation } from '../utils/dom/event';

const Ellipsis = props => {
  const rootRef = useRef(null);
  const [ellipsised, setEllipsised] = useState({});
  const [expanded, setExpanded] = useState(false);
  const [exceeded, setExceeded] = useState(false);
  const suffixStr = useMemo(() => {
    if (props.suffixText) return props.suffixText;

    if (props.suffixCount > 0) {
      return props.children.slice(-props.suffixCount).trim();
    }

    return '';
  }, [props.suffixCount, props.suffixText]);

  function calcEllipsised() {
    const root = rootRef.current;
    if (!root) return;
    const originStyle = window.getComputedStyle(root);
    const container = document.createElement('div');
    const styleNames = Array.prototype.slice.apply(originStyle);
    styleNames.forEach(name => {
      container.style.setProperty(name, originStyle.getPropertyValue(name));
    });
    container.style.position = 'fixed';
    container.style.left = '999999px';
    container.style.top = '999999px';
    container.style.zIndex = '-1000';
    container.style.height = 'auto';
    container.style.minHeight = 'auto';
    container.style.maxHeight = 'auto';
    container.style.textOverflow = 'clip';
    container.style.whiteSpace = 'normal';
    container.style.webkitLineClamp = 'unset';
    container.style.display = 'block';
    const lineHeight = pxToNumber(originStyle.lineHeight);
    const maxHeight = Math.floor(lineHeight * (props.rows + 0.5) + pxToNumber(originStyle.paddingTop) + pxToNumber(originStyle.paddingBottom));
    container.innerHTML = props.children;
    document.body.appendChild(container);

    if (container.offsetHeight <= maxHeight) {
      setExceeded(false);
    } else {
      setExceeded(true);
      const end = props.children.length;
      const actionText = expanded ? props.collapseText : props.expandText; // eslint-disable-next-line no-inner-declarations

      function check(left, right) {
        if (right - left <= 1) {
          return {
            leading: props.children.slice(0, left) + props.symbol
          };
        }

        const middle = Math.round((left + right) / 2);
        container.innerHTML = props.children.slice(0, middle) + props.symbol + suffixStr + actionText;

        if (container.offsetHeight <= maxHeight) {
          return check(middle, right);
        }

        return check(left, middle);
      }

      const ellipsised = check(0, end);
      setEllipsised(ellipsised);
    }

    document.body.removeChild(container);
  }

  useResizeEffect(calcEllipsised, rootRef);
  useIsomorphicLayoutEffect(() => {
    calcEllipsised();
  }, [props.children, props.rows, suffixStr, props.expandText, props.collapseText]);

  const onExpandClick = isExpend => {
    var _a;

    (_a = props.onExpend) === null || _a === void 0 ? void 0 : _a.call(props, isExpend);
    setExpanded(isExpend);
  };

  const expandActionElement = exceeded && props.expandText ? withStopPropagation(props.stopPropagationForActionButtons, _jsx("a", Object.assign({
    onClick: () => {
      onExpandClick(true);
    }
  }, {
    children: props.expandText
  }))) : null;
  const collapseActionElement = exceeded && props.expandText ? withStopPropagation(props.stopPropagationForActionButtons, _jsx("a", Object.assign({
    onClick: () => {
      onExpandClick(false);
    }
  }, {
    children: props.collapseText
  }))) : null;

  const renderContent = () => {
    if (!exceeded) {
      return props.children;
    }

    if (expanded) {
      return _jsxs(_Fragment, {
        children: [props.children, collapseActionElement]
      });
    } else {
      return _jsxs(_Fragment, {
        children: [ellipsised.leading, suffixStr, expandActionElement]
      });
    }
  };

  return _jsx("div", Object.assign({
    ref: rootRef,
    className: clsx('rv-typography__ellipsis', props.className),
    style: props.style
  }, {
    children: renderContent()
  }));
};

function pxToNumber(value) {
  if (!value) return 0;
  const match = value.match(/^\d*(\.\d*)?/);
  return match ? Number(match[0]) : 0;
}

Ellipsis.defaultProps = {
  rows: 1,
  expandText: '',
  collapseText: '',
  suffixText: '',
  symbol: '...',
  stopPropagationForActionButtons: []
};
export default Ellipsis;
