// react-vant/lib/cell/Cell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

function _icons() {
  const data = require("@react-vant/icons");

  _icons = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('cell');

const Cell = props => {
  const renderLabel = () => {
    const showLabel = (0, _utils.isDef)(props.label);

    if (showLabel) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('label'), props.labelClass)
      }, {
        children: props.label
      }));
    }

    return null;
  };

  const renderTitle = () => {
    if ((0, _utils.isDef)(props.title)) {
      return (0, _jsxRuntime().jsxs)("div", Object.assign({
        className: (0, _clsx().default)(bem('title'), props.titleClass),
        style: props.titleStyle
      }, {
        children: [props.title, renderLabel()]
      }));
    }

    return null;
  };

  const renderValue = () => {
    const hasTitle = (0, _utils.isDef)(props.title);
    const hasValue = props.children || (0, _utils.isDef)(props.value);

    if (hasValue) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('value', {
          alone: !hasTitle
        }), props.valueClass)
      }, {
        children: props.children ? props.children : (0, _jsxRuntime().jsx)("span", {
          children: props.value
        })
      }));
    }

    return null;
  };

  const renderLeftIcon = () => {
    if (props.icon) {
      return _react().default.cloneElement(props.icon, {
        className: (0, _clsx().default)(bem('left-icon'))
      });
    }

    return null;
  };

  const renderRightIcon = () => {
    if (props.rightIcon) {
      return props.rightIcon;
    }

    if (props.isLink) {
      const className = (0, _clsx().default)(bem('right-icon'));
      if (props.arrowDirection === 'left') return (0, _jsxRuntime().jsx)(_icons().ArrowLeft, {
        className: className
      });
      if (props.arrowDirection === 'up') return (0, _jsxRuntime().jsx)(_icons().ArrowUp, {
        className: className
      });
      if (props.arrowDirection === 'down') return (0, _jsxRuntime().jsx)(_icons().ArrowDown, {
        className: className
      });
      return (0, _jsxRuntime().jsx)(_icons().Arrow, {
        className: className
      });
    }

    return null;
  };

  const {
    className,
    style,
    size,
    center,
    border = true,
    isLink,
    required,
    onClick
  } = props;
  const clickable = isLink || props.clickable;
  const classes = {
    center,
    required,
    clickable,
    borderless: !border
  };

  if (size) {
    classes[size] = !!size;
  }

  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    style: style,
    className: (0, _clsx().default)(bem(classes), className),
    role: clickable ? 'button' : undefined,
    onClick: onClick
  }, {
    children: [renderLeftIcon(), renderTitle(), renderValue(), renderRightIcon(), props.extra]
  }));
};

var _default = Cell;
exports.default = _default;
