// react-vant/lib/selector/Selector.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Selector = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _hooks = require("../hooks");

var _CheckMark = require("./CheckMark");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// https://mobile.ant.design/zh/components/selector
const [bem] = (0, _utils.createNamespace)('selector');
const defaultProps = {
  multiple: false,
  defaultValue: [],
  showCheckMark: true
};

const Selector = p => {
  const props = Object.assign(Object.assign({}, defaultProps), p);
  const [value, setValue] = (0, _hooks.usePropsValue)({
    value: props.value,
    defaultValue: props.defaultValue,
    onChange: val => {
      var _a;

      const extend = {
        get items() {
          return props.options.filter(option => val.includes(option.value));
        }

      };
      (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, val, extend);
    }
  });
  const items = props.options.map(option => {
    const active = (value || []).includes(option.value);
    const disabled = option.disabled || props.disabled;
    const itemCls = (0, _clsx().default)(bem('item', {
      active: active && !props.multiple,
      'multiple-active': active && props.multiple,
      disabled
    }));
    return (0, _jsxRuntime().jsxs)("div", Object.assign({
      className: itemCls,
      onClick: () => {
        if (disabled) {
          return;
        }

        if (props.multiple) {
          const val = active ? value.filter(v => v !== option.value) : [...value, option.value];
          setValue(val);
        } else {
          const val = active ? [] : [option.value];
          setValue(val);
        }
      }
    }, {
      children: [option.label, option.description && (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('item-description'))
      }, {
        children: option.description
      })), active && props.showCheckMark && (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('check-mark-wrapper'))
      }, {
        children: (0, _jsxRuntime().jsx)(_CheckMark.CheckMark, {})
      }))]
    }), option.value);
  });
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(bem(), props.className),
    style: props.style
  }, {
    children: items
  }));
};

exports.Selector = Selector;
