// react-vant/lib/button/ButtonGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ButtonGroup = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _ButtonContext = _interopRequireDefault(require("./ButtonContext"));

var _constant = require("../utils/constant");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('button-group');

const ButtonGroup = _a => {
  var {
    className,
    style,
    children,
    onClick
  } = _a,
      props = (0, _tslib().__rest)(_a, ["className", "style", "children", "onClick"]);

  const internalClick = e => {
    if (props.disabled) return;
    onClick === null || onClick === void 0 ? void 0 : onClick(e);
  };

  return (0, _jsxRuntime().jsx)("div", Object.assign({
    onClick: internalClick,
    style: style,
    className: (0, _clsx().default)(className, bem([props.type, {
      round: props.round,
      plain: props.plain,
      disabled: props.disabled
    }]), props.shadow && `${_constant.SHADOW}--${+props.shadow}`)
  }, {
    children: (0, _jsxRuntime().jsx)(_ButtonContext.default.Provider, Object.assign({
      value: {
        parent: props
      }
    }, {
      children: children
    }))
  }));
};

exports.ButtonGroup = ButtonGroup;
var _default = ButtonGroup;
exports.default = _default;
