// react-vant/lib/swiper/SwiperItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

var _hooks = require("../hooks");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('swiper-item');

const SwiperItem = _react().default.forwardRef((props, ref) => {
  const wrapperRef = _react().default.useRef(null);

  _react().default.useImperativeHandle(ref, () => ({
    self: wrapperRef.current
  }));

  const [show] = (0, _hooks.useInViewport)(wrapperRef, {
    rootMargin: '-0.1px',
    threshold: 0,
    root: () => {
      var _a;

      return (_a = props.trackRef) === null || _a === void 0 ? void 0 : _a.current;
    }
  });
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    ref: wrapperRef,
    className: (0, _clsx().default)(props.className, bem({
      hidden: props.autoHeight && show === false
    })),
    onClick: props.onClick,
    style: props.style
  }, {
    children: props.children
  }));
});

var _default = SwiperItem;
exports.default = _default;
