// react-vant/es/cascader/Cascader.js
import { __rest } from "tslib";
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
/* eslint-disable no-plusplus */

import React, { useContext, useEffect, useImperativeHandle, useState } from 'react';
import cls from 'clsx';
import { Cross, Success } from '@react-vant/icons';
import { createNamespace, extend, isObject } from '../utils';
import { useMemoizedFn, usePropsValue, useUpdateEffect } from '../hooks';
import Tabs from '../tabs';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import Popup from '../popup';
import { useCascaderExtend } from './useCascaderExtend';
import useDebounceEffect from '../hooks/use-debunce-effect';
const [bem] = createNamespace('cascader');

const Cascader = props => {
  const {
    locale
  } = useContext(ConfigProviderContext);
  const [value, setValue] = useState(() => props.value === undefined ? props.defaultValue : props.value);
  const [activeTab, updateActiveTab] = useState(0);
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = extend({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.fieldNames);
  const {
    tabs,
    items,
    depth
  } = useCascaderExtend(props.options, {
    textKey,
    valueKey,
    childrenKey
  }, value || []); // sync props.value to inner value

  useUpdateEffect(() => {
    if (props.value === undefined) return; // uncontroll mode

    if (JSON.stringify(value) !== JSON.stringify(props.value)) {
      setValue(props.value);
    }
  }, [props.value]); // update active tab index from value

  useEffect(() => {
    let tabIndex = 0;
    if (Array.isArray(value) && value.length > 0) tabIndex = value.length;
    if (tabIndex >= depth) tabIndex = depth - 1;
    if (tabIndex === activeTab) return;
    updateActiveTab(tabIndex);
  }, [value]);
  useDebounceEffect(() => {
    var _a, _b;

    if (JSON.stringify(props.value) === JSON.stringify(value)) return;
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, value, items);

    if (value.length >= depth) {
      (_b = props.onFinish) === null || _b === void 0 ? void 0 : _b.call(props, value, items);
    }
  }, [value], {
    wait: 0,
    leading: false,
    trailing: true
  });

  const onSelect = (option, tabIndex) => {
    if (option.disabled) {
      return;
    }

    setValue(prev => {
      const next = [...prev];
      next[tabIndex] = option === null || option === void 0 ? void 0 : option[valueKey];
      return next.slice(0, tabIndex + 1);
    });
  };

  const onClickTab = ({
    name,
    index
  }) => {
    var _a;

    (_a = props.onClickTab) === null || _a === void 0 ? void 0 : _a.call(props, +name);
    updateActiveTab(index);
  };

  const renderCloseIcon = () => {
    if (!props.closeable) return null;

    if (props.closeIcon) {
      return React.cloneElement(props.closeIcon, {
        className: cls(bem('close-icon')),
        onClick: props.onClose
      });
    }

    return _jsx(Cross, {
      className: cls(bem('close-icon')),
      onClick: props.onClose
    });
  };

  const renderHeader = () => _jsxs("div", Object.assign({
    className: cls(bem('header'))
  }, {
    children: [_jsx("h2", Object.assign({
      className: cls(bem('title'))
    }, {
      children: props.title
    })), renderCloseIcon()]
  }));

  const renderOption = (option, selected, tabIndex) => {
    const color = option.color || (selected ? props.activeColor : undefined);
    const Text = props.optionRender ? props.optionRender({
      option,
      selected
    }) : _jsx("span", {
      children: option[textKey]
    });
    return _jsxs("li", Object.assign({
      className: cls(bem('option', {
        selected,
        disabled: option.disabled
      }), option.className),
      style: {
        color
      },
      onClick: () => onSelect(option, tabIndex)
    }, {
      children: [Text, selected ? _jsx(Success, {
        className: cls(bem('selected-icon'))
      }) : null]
    }), option[valueKey]);
  };

  const renderOptions = (options, selectedOption, tabIndex) => _jsx("ul", Object.assign({
    className: cls(bem('options'))
  }, {
    children: options.map(option => renderOption(option, option[valueKey] === (selectedOption === null || selectedOption === void 0 ? void 0 : selectedOption[valueKey]), tabIndex))
  }), tabIndex);

  const renderTab = (options, tabIndex) => {
    const selectedOption = items[tabIndex];
    const title = selectedOption ? selectedOption[textKey] : props.placeholder || locale.vanPicker.select;
    return _jsx(Tabs.TabPane, Object.assign({
      title: title,
      titleClass: cls(bem('tab', {
        unselected: !selectedOption
      }))
    }, {
      children: renderOptions(options, selectedOption, tabIndex)
    }), tabIndex);
  };

  const renderTabs = () => _jsx(Tabs, Object.assign({
    animated: true,
    active: activeTab,
    className: cls(bem('tabs')),
    color: props.activeColor,
    swipeThreshold: 0,
    swipeable: props.swipeable,
    duration: 300,
    onClickTab: onClickTab
  }, {
    children: tabs.map(renderTab)
  }));

  return _jsxs("div", Object.assign({
    className: cls(bem())
  }, {
    children: [renderHeader(), tabs.length ? renderTabs() : null]
  }));
};

const CascaderPopup = React.forwardRef((props, ref) => {
  var _a;

  const {
    visible: outerVisible,
    popup
  } = props,
        cascaderProps = __rest(props, ["visible", "popup"]);

  const [visible, setVisible] = usePropsValue({
    value: outerVisible,
    defaultValue: false
  });
  const [value, setValue] = useState(() => props.value === undefined ? props.defaultValue : props.value);
  const [innerValue, setInnerValue] = useState(value);
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = extend({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.fieldNames);
  const {
    items
  } = useCascaderExtend(props.options, {
    textKey,
    valueKey,
    childrenKey
  }, value || []); // sync props.value to inner value

  useUpdateEffect(() => {
    if (props.value === undefined) return; // uncontroll mode

    if (JSON.stringify(value) !== JSON.stringify(props.value)) {
      setValue(props.value);
    }
  }, [props.value]); // sync value to cascader value

  useEffect(() => {
    if (!popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [value]); // only sync value to popup cascader value when visible change

  useEffect(() => {
    if (popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [visible]);
  const actions = {
    toggle: () => {
      if (popup) setVisible(v => !v);
    },
    open: () => {
      if (popup) {
        setVisible(true);
      }
    },
    close: () => {
      if (popup) {
        setVisible(false);
      }
    }
  };

  const onClose = () => {
    var _a;

    (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
    actions.close();
  };

  const onFinish = useMemoizedFn((val, selectedRows) => {
    var _a;

    setValue(val);
    (_a = props.onFinish) === null || _a === void 0 ? void 0 : _a.call(props, val, selectedRows);
    actions.close();
  });
  const onChange = useMemoizedFn((val, selectedRows) => {
    var _a, _b;

    setInnerValue(val);

    if (popup) {
      if (visible) (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, val, selectedRows);
    } else {
      (_b = props.onChange) === null || _b === void 0 ? void 0 : _b.call(props, val, selectedRows);
    }
  });
  useImperativeHandle(ref, () => actions);

  const content = _jsx(Cascader, Object.assign({
    value: innerValue
  }, cascaderProps, {
    onChange: onChange,
    onFinish: onFinish,
    onClose: onClose
  }));

  if (!popup) return content;
  const popupProps = isObject(popup) ? Object.assign({
    closeOnClickOverlay: true
  }, popup) : {
    closeOnClickOverlay: true
  };
  return _jsxs(_Fragment, {
    children: [_jsx(Popup, Object.assign({
      position: 'bottom',
      visible: visible,
      closeOnClickOverlay: true,
      onClickOverlay: () => {
        if (!(popupProps === null || popupProps === void 0 ? void 0 : popupProps.closeOnClickOverlay)) return;
        setVisible(false);
      }
    }, popupProps, {
      children: content
    })), (_a = props.children) === null || _a === void 0 ? void 0 : _a.call(props, value, items, actions)]
  });
});
CascaderPopup.defaultProps = {
  closeable: true,
  swipeable: false,
  defaultValue: [],
  options: []
};
export default CascaderPopup;
