// react-vant/es/datetime-picker/TimePicker.js
import { __rest } from "tslib";
import { jsx as _jsx } from "react/jsx-runtime";
import { forwardRef, useMemo, useState } from 'react';
import Picker from '../picker';
import { times } from './utils';
import { range, padZero } from '../utils';
import { useUpdateEffect } from '../hooks';
const TimePicker = forwardRef((props, ref) => {
  const {
    value,
    defaultValue,
    formatter,
    filter,
    minHour,
    maxHour,
    minMinute,
    maxMinute
  } = props,
        pickerProps = __rest(props, ["value", "defaultValue", "formatter", "filter", "minHour", "maxHour", "minMinute", "maxMinute"]);

  const formatValue = str => {
    if (!str) {
      str = `${padZero(minHour)}:${padZero(minMinute)}`;
    }

    let [hour, minute] = str.split(':');
    hour = padZero(range(hour, +minHour, +maxHour));
    minute = padZero(range(minute, +minMinute, +maxMinute));
    return `${hour}:${minute}`;
  };

  const [currentDate, setCurrentDate] = useState(() => formatValue(value === undefined ? defaultValue : value));
  const ranges = useMemo(() => [{
    type: 'hour',
    range: [+minHour, +maxHour]
  }, {
    type: 'minute',
    range: [+minMinute, +maxMinute]
  }], [minHour, maxHour, minMinute, maxMinute]);
  const originColumns = useMemo(() => ranges.map(({
    type,
    range: rangeArr
  }) => {
    let values = times(rangeArr[1] - rangeArr[0] + 1, index => padZero(rangeArr[0] + index));

    if (filter) {
      values = filter(type, values);
    }

    return {
      type,
      values
    };
  }), [ranges]);
  const columns = useMemo(() => originColumns.map(column => column.values.map(value => formatter(column.type, value))), [originColumns]);
  const pickerValue = useMemo(() => {
    const pair = (props.popup ? formatValue(props.value) : currentDate).split(':');
    return [formatter('hour', pair[0]), formatter('minute', pair[1])];
  }, [props.value, currentDate, formatValue]);

  const onConfirm = () => {
    var _a;

    (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props, currentDate);
  };

  const onCancel = () => {
    var _a;

    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(props);
  };

  const onChange = val => {
    var _a;

    const nextValue = formatValue(val.join(':'));
    setCurrentDate(nextValue);
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, nextValue);
  };

  useUpdateEffect(() => {
    const nextValue = formatValue(currentDate);
    setCurrentDate(nextValue);
  }, [filter, minHour, maxHour, minMinute, maxMinute]);
  useUpdateEffect(() => {
    const nextValue = formatValue(value);

    if (nextValue !== currentDate) {
      setCurrentDate(nextValue);
    }
  }, [value]);
  return _jsx(Picker, Object.assign({}, pickerProps, {
    ref: ref,
    columns: columns,
    value: pickerValue,
    onChange: onChange,
    onConfirm: onConfirm,
    onCancel: onCancel
  }, {
    children: (_, selectRows, actions) => {
      var _a;

      return (_a = props.children) === null || _a === void 0 ? void 0 : _a.call(props, value, selectRows, actions);
    }
  }));
});
TimePicker.defaultProps = {
  minHour: 0,
  maxHour: 23,
  minMinute: 0,
  maxMinute: 59,
  placeholder: false,
  defaultValue: '',
  formatter: (type, value) => value
};
export default TimePicker;
