// react-vant/lib/tabs/Tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _useRefs = _interopRequireDefault(require("../hooks/use-refs"));

var _useScrollParent = _interopRequireDefault(require("../hooks/use-scroll-parent"));

var _sticky = _interopRequireDefault(require("../sticky"));

var _TabsTitle = _interopRequireDefault(require("./TabsTitle"));

var _TabsContent = _interopRequireDefault(require("./TabsContent"));

var _TabsContext = _interopRequireDefault(require("./TabsContext"));

var _utils = require("../utils");

var _interceptor = require("../utils/interceptor");

var _constant = require("../utils/constant");

var _hooks = require("../hooks");

var _useEventListener = _interopRequireDefault(require("../hooks/use-event-listener"));

var _utils2 = require("./utils");

var _useRefState = _interopRequireDefault(require("../hooks/use-ref-state"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const [bem] = (0, _utils.createNamespace)('tabs');

const getTabName = (tab, index) => {
  var _a;

  return (_a = tab === null || tab === void 0 ? void 0 : tab.name) !== null && _a !== void 0 ? _a : index;
};

const Tabs = (0, _react().forwardRef)((props, ref) => {
  const {
    children,
    color,
    align,
    background
  } = props;
  const root = (0, _react().useRef)(null);
  const swiperRef = (0, _react().useRef)(null);
  const wrapRef = (0, _react().useRef)(null);
  const lockScroll = (0, _react().useRef)(false);
  const stickyFixed = (0, _react().useRef)(false);
  const immediateRef = (0, _react().useRef)(true);
  const navRef = (0, _react().useRef)(null);
  const scroller = (0, _useScrollParent.default)(root);
  const [titleRefs, setTitleRefs] = (0, _useRefs.default)();
  const [contentRefs, setContentRefs] = (0, _useRefs.default)();
  const wrapHeight = (0, _react().useMemo)(() => {
    if (!wrapRef.current) return 0;
    return (0, _utils.getVisibleHeight)(wrapRef.current);
  }, [wrapRef.current]);
  const childrenList = (0, _react().useMemo)(() => (0, _utils.parseChildList)(props.children), [props.children]);
  const defaultIndex = (0, _react().useMemo)(() => {
    if (props.scrollspy) return 0;
    const ac = props.active === undefined ? props.defaultActive : props.active;
    let idx = childrenList.findIndex((tab, index) => getTabName(tab, index) === ac);
    if (idx < 0) idx = 0;
    return idx;
  }, [props.scrollspy, props.active, props.defaultActive]);
  const [index, setIndex] = (0, _useRefState.default)(defaultIndex); // whether the nav is scrollable

  const scrollable = (0, _react().useMemo)(() => childrenList.length > props.swipeThreshold || !props.ellipsis, [childrenList.length, props.swipeThreshold, props.ellipsis]);
  const navStyle = (0, _react().useMemo)(() => ({
    borderColor: props.type === 'card' && color,
    background
  }), [color, background]);
  const currentName = (0, _react().useMemo)(() => {
    const activeTab = childrenList === null || childrenList === void 0 ? void 0 : childrenList[index];
    return activeTab ? getTabName(activeTab, index) : 0;
  }, [childrenList, index]);
  const offsetTopPx = (0, _react().useMemo)(() => (0, _utils.unitToPx)(props.offsetTop), [props.offsetTop]); // 下划线偏移量

  const lineTranslateLeft = (0, _react().useMemo)(() => {
    const hidden = (0, _utils.isHidden)(root.current);
    const title = titleRefs === null || titleRefs === void 0 ? void 0 : titleRefs[index];

    if (!title || hidden || props.type !== 'line') {
      return;
    }

    return title.offsetLeft + title.offsetWidth / 2;
  }, [root.current, titleRefs, props.type, index]); // 下划线样式

  const lineStyle = (0, _react().useMemo)(() => {
    const {
      lineWidth,
      lineHeight
    } = props;
    const measureStyle = {
      width: (0, _utils.addUnit)(lineWidth),
      backgroundColor: color,
      transitionDuration: `${immediateRef.current ? 0 : props.duration}ms`
    };

    if (lineTranslateLeft) {
      measureStyle.transform = `translateX(${lineTranslateLeft}px) translateX(-50%)`;
    }

    if (lineHeight) {
      const height = (0, _utils.addUnit)(lineHeight);
      measureStyle.height = height;
      measureStyle.borderRadius = height;
    }

    return measureStyle;
  }, [color, props.lineHeight, props.lineWidth, lineTranslateLeft, immediateRef.current]);

  const getAvailableTab = targetIndex => {
    var _a;

    const diff = targetIndex < index ? -1 : 1;

    while (targetIndex >= 0 && targetIndex < childrenList.length) {
      if (!((_a = childrenList[targetIndex]) === null || _a === void 0 ? void 0 : _a.disabled)) {
        return targetIndex;
      }

      targetIndex += diff;
    }

    return null;
  };

  const setCurrentIndex = currentIndex => {
    var _a;

    const availableIndex = getAvailableTab(currentIndex);
    if (availableIndex === null) return;
    const newTab = childrenList[availableIndex];
    const newName = getTabName(newTab, availableIndex);
    setIndex(availableIndex);
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, newName, availableIndex);
  }; // sync scrollspy content position


  const scrollToCurrentContent = current => {
    var _a, _b;

    if (!props.scrollspy) return;
    const contentTarget = contentRefs[current !== null && current !== void 0 ? current : index];

    if (contentTarget && scroller) {
      const immediate = (0, _utils.isObject)(props.scrollspy) ? (_b = (_a = props.scrollspy) === null || _a === void 0 ? void 0 : _a.scrollImmediate) !== null && _b !== void 0 ? _b : true : true; // void ui shake use math.ceil to get `to` value

      const to = Math.ceil((0, _utils.getElementTop)(contentTarget, scroller) - (offsetTopPx + wrapHeight));
      lockScroll.current = true;
      (0, _utils.scrollTopTo)(scroller, to, immediate ? 0 : +props.duration, () => {
        lockScroll.current = false;
      });
    }
  };

  const onClickTab = (item, targetIndex, event) => {
    var _a;

    const {
      disabled = false
    } = item;
    const name = getTabName(item, targetIndex);
    (_a = props.onClickTab) === null || _a === void 0 ? void 0 : _a.call(props, {
      name,
      event,
      disabled,
      index: targetIndex
    });
    if (disabled) return;
    (0, _interceptor.callInterceptor)({
      interceptor: props.beforeChange,
      args: [name],
      done: () => {
        if (targetIndex !== index) {
          setCurrentIndex(targetIndex);
          scrollToCurrentContent(targetIndex);
        }
      }
    });
  }; // scroll active tab into view


  const scrollIntoView = immediate => {
    const nav = navRef.current;
    const title = titleRefs === null || titleRefs === void 0 ? void 0 : titleRefs[index];

    if (!scrollable || !nav || !title) {
      return;
    }

    const to = title.offsetLeft - (nav.offsetWidth - title.offsetWidth) / 2;
    (0, _utils.scrollLeftTo)(nav, to, immediate ? 0 : +props.duration);
  };

  const getCurrentIndexOnScroll = () => {
    const scrollOffset = offsetTopPx + wrapHeight; // eslint-disable-next-line no-plusplus

    for (let index = 0; index < contentRefs.length; index++) {
      const $el = contentRefs[index];
      const top = (0, _utils.getVisibleTop)($el);

      if (top > scrollOffset) {
        return index === 0 ? 0 : index - 1;
      }
    }

    return titleRefs.length - 1;
  };

  const onScroll = () => {
    if (!props.scrollspy || lockScroll.current) return;
    let currentIndex = getCurrentIndexOnScroll();

    if (typeof props.scrollspy === 'object') {
      if (props.scrollspy.autoFocusLast && (0, _utils2.isReachBottom)(props.scrollspy.reachBottomThreshold)) {
        currentIndex = titleRefs.length - 1;
      }
    }

    if (currentIndex !== index) {
      setCurrentIndex(currentIndex);
    }
  };

  const onStickyScroll = params => {
    var _a;

    stickyFixed.current = params.isFixed;
    (_a = props.onScroll) === null || _a === void 0 ? void 0 : _a.call(props, params);
  };

  const renderNav = () => {
    return childrenList.map((item, itemIndex) => {
      return (0, _jsxRuntime().jsx)(_TabsTitle.default, {
        ref: setTitleRefs(itemIndex),
        type: props.type,
        badge: item.badge,
        title: item.title,
        description: item.description,
        color: color,
        style: item.titleStyle,
        className: item.titleClass,
        isActive: itemIndex === index,
        disabled: item.disabled,
        scrollable: scrollable,
        activeColor: props.titleActiveColor,
        inactiveColor: props.titleInactiveColor,
        onClick: event => {
          onClickTab(item, itemIndex, event);
        }
      }, item.key);
    });
  };

  const renderHeader = () => {
    const {
      type,
      border
    } = props;
    return (0, _jsxRuntime().jsx)("div", Object.assign({
      ref: wrapRef,
      className: (0, _clsx().default)([bem('wrap', {
        scrollable
      }), {
        [_constant.BORDER_TOP_BOTTOM]: type !== 'card' && border
      }])
    }, {
      children: (0, _jsxRuntime().jsxs)("div", Object.assign({
        ref: navRef,
        role: 'tablist',
        className: (0, _clsx().default)(bem('nav', [type, {
          complete: scrollable,
          start: align === 'start'
        }])),
        style: navStyle
      }, {
        children: [props.navLeft, renderNav(), type === 'line' && !immediateRef.current && (0, _jsxRuntime().jsx)("div", {
          className: (0, _clsx().default)(bem('line')),
          style: lineStyle
        }), props.navRight]
      }))
    }));
  }; // sync props.active to inner index value


  (0, _hooks.useUpdateEffect)(() => {
    if (props.active === undefined) return;

    if (props.active !== currentName) {
      const currentIndex = childrenList.findIndex((tab, index) => getTabName(tab, index) === props.active);

      if (currentIndex > -1 && currentIndex !== index) {
        setIndex(currentIndex);
        scrollToCurrentContent(currentIndex);
      }
    }
  }, [props.active]); // sync tab.pane title in correct postion when index change

  (0, _hooks.useUpdateEffect)(() => {
    scrollIntoView(); // scrollspy situation

    if (stickyFixed.current && props.stickyInitScrollbar && !props.scrollspy) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      (0, _utils.setRootScrollTop)(Math.ceil((0, _utils.getElementTop)(root.current) - offsetTopPx));
    }
  }, [index]); // init

  (0, _react().useEffect)(() => {
    immediateRef.current = false;
    scrollIntoView(true);
  }, []);
  (0, _useEventListener.default)('scroll', onScroll, {
    target: scroller,
    depends: [index]
  });
  (0, _react().useImperativeHandle)(ref, () => {
    var _a, _b;

    return {
      scrollTo: name => {
        const currentIndex = childrenList.findIndex((tab, index) => getTabName(tab, index) === name);

        if (currentIndex > -1 && currentIndex !== index) {
          setIndex(currentIndex);
          scrollToCurrentContent(currentIndex);
        }
      },
      swiper: swiperRef.current ? {
        enable: (_a = swiperRef.current) === null || _a === void 0 ? void 0 : _a.enable,
        disable: (_b = swiperRef.current) === null || _b === void 0 ? void 0 : _b.disable
      } : undefined
    };
  });
  return (0, _jsxRuntime().jsx)(_TabsContext.default.Provider, Object.assign({
    value: {
      props,
      currentName,
      scrollIntoView
    }
  }, {
    children: (0, _jsxRuntime().jsxs)("div", Object.assign({
      ref: root,
      className: (0, _clsx().default)(props.className, bem([props.type]))
    }, {
      children: [props.sticky ? (0, _jsxRuntime().jsxs)(_sticky.default, Object.assign({
        container: root,
        offsetTop: offsetTopPx,
        onScroll: onStickyScroll
      }, {
        children: [renderHeader(), props.navBottom]
      })) : (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
        children: [renderHeader(), props.navBottom]
      }), (0, _jsxRuntime().jsx)(_TabsContent.default, Object.assign({
        swiperRef: swiperRef,
        count: childrenList.length,
        animated: props.animated,
        duration: props.duration,
        swipeable: props.swipeable,
        lazyRender: props.lazyRender,
        currentIndex: index,
        onChange: setCurrentIndex
      }, {
        children: _react().default.Children.toArray(children).filter(Boolean).map((node, index) => _react().default.cloneElement(node, {
          index,
          ref: setContentRefs(index)
        }))
      }))]
    }))
  }));
});
Tabs.defaultProps = {
  type: 'line',
  duration: 300,
  swipeThreshold: 5,
  offsetTop: 0,
  ellipsis: true,
  lazyRender: true,
  stickyInitScrollbar: true,
  defaultActive: 0,
  align: 'center'
};
var _default = Tabs;
exports.default = _default;
