// react-vant/lib/skeleton/Skeleton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_ROW_WIDTH = '100%';
const DEFAULT_LAST_ROW_WIDTH = '60%';
const [bem] = (0, _utils.createNamespace)('skeleton');

const Skeleton = _a => {
  var {
    children,
    className,
    style
  } = _a,
      props = (0, _tslib().__rest)(_a, ["children", "className", "style"]);

  const getRowWidth = index => {
    const {
      rowWidth
    } = props;

    if (rowWidth === DEFAULT_ROW_WIDTH && index === +props.row - 1 && index !== 0) {
      return DEFAULT_LAST_ROW_WIDTH;
    }

    if (Array.isArray(rowWidth)) {
      return rowWidth[index];
    }

    return rowWidth;
  };

  const getRowHeight = index => {
    const {
      rowHeight
    } = props;

    if (Array.isArray(rowHeight)) {
      return rowHeight[index];
    }

    return rowHeight;
  };

  const renderAvatar = () => {
    if (props.avatar) {
      return (0, _jsxRuntime().jsx)("div", {
        className: (0, _clsx().default)(bem('avatar', props.avatarShape)),
        style: (0, _utils.getSizeStyle)(props.avatarSize)
      });
    }

    return null;
  };

  const renderTitle = () => {
    if (props.title) {
      const width = (0, _utils.addUnit)(props.titleWidth);
      const height = (0, _utils.addUnit)(getRowHeight(0));
      return (0, _jsxRuntime().jsx)("div", {
        className: (0, _clsx().default)(bem('title')),
        style: {
          width,
          height
        }
      });
    }

    return null;
  };

  const renderRows = () => Array(props.row).fill('').map((_, i) => {
    const width = (0, _utils.addUnit)(getRowWidth(i));
    const height = (0, _utils.addUnit)(getRowHeight(i)); // eslint-disable-next-line react/no-array-index-key

    return (0, _jsxRuntime().jsx)("div", {
      className: (0, _clsx().default)(bem('row')),
      style: {
        width,
        height
      }
    }, i);
  });

  if (!props.loading) return (0, _jsxRuntime().jsx)(_jsxRuntime().Fragment, {
    children: children
  });
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(className, bem({
      animate: props.animate,
      round: props.round
    })),
    style: style
  }, {
    children: [renderAvatar(), (0, _jsxRuntime().jsxs)("div", Object.assign({
      className: (0, _clsx().default)(bem('content'))
    }, {
      children: [renderTitle(), renderRows()]
    }))]
  }));
};

Skeleton.defaultProps = {
  loading: true,
  animate: true,
  round: true,
  row: 3,
  avatarShape: 'round',
  rowWidth: DEFAULT_ROW_WIDTH
};
var _default = Skeleton;
exports.default = _default;
