// react-vant/lib/radio/RadioGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _useMergedState = _interopRequireDefault(require("../hooks/use-merged-state"));

var _RadioContext = _interopRequireDefault(require("./RadioContext"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('radio-group');

function RadioGroup(props) {
  const [checked, setChecked] = (0, _useMergedState.default)({
    value: props.value,
    defaultValue: props.defaultValue
  });

  const toggle = name => {
    var _a;

    setChecked(name);
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, name);
  };

  return (0, _jsxRuntime().jsx)(_RadioContext.default.Provider, Object.assign({
    value: {
      parent: {
        props
      },
      toggle,
      checked
    }
  }, {
    children: (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(props.className, bem([props.direction])),
      style: props.style,
      role: 'radiogroup'
    }, {
      children: props.children
    }))
  }));
}

var _default = RadioGroup;
exports.default = _default;
