// react-vant/es/tabs/TabsContent.js
import { jsx as _jsx } from "react/jsx-runtime";
import React, { useRef, useEffect } from 'react';
import clsx from 'clsx';
import Swiper from '../swiper';
import { createNamespace } from '../utils';
const [bem] = createNamespace('tabs');

const TabsContent = props => {
  const innerEffect = useRef(false);
  const {
    animated,
    swipeable,
    duration,
    swiperRef
  } = props;
  const swiperProps = typeof swipeable === 'boolean' ? {} : swipeable;

  const renderChildren = () => {
    if (animated || swipeable) {
      return _jsx(Swiper, Object.assign({}, swiperProps, {
        ref: swiperRef,
        rubberband: false,
        stuckAtBoundary: true,
        loop: false,
        autoplay: false,
        touchable: !!swipeable,
        className: clsx(bem('track')),
        duration: +duration,
        indicator: false,
        onChange: idx => {
          var _a;

          if (innerEffect.current) {
            innerEffect.current = false;
            return;
          }

          (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, idx);
        }
      }, {
        children: React.Children.map(props.children, child => _jsx(Swiper.Item, Object.assign({
          style: {
            cursor: !swipeable ? 'auto' : undefined
          },
          className: clsx(bem('pane-wrapper'))
        }, {
          children: child
        })))
      }));
    }

    return props.children;
  };

  const swipeToCurrentTab = index => {
    const swipe = swiperRef.current;
    if (!swipe) return;

    if (swipe.activeIndex !== index) {
      innerEffect.current = true;
      swipe.swipeTo(index);
    }
  };

  useEffect(() => {
    swipeToCurrentTab(props.currentIndex);
  }, [props.currentIndex]);
  return _jsx("div", Object.assign({
    className: clsx(bem('content', {
      animated: animated || swipeable
    }))
  }, {
    children: renderChildren()
  }));
};

export default TabsContent;
