// react-vant/es/tabs/Tabs.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import React, { useMemo, useRef, useEffect, forwardRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import useRefs from '../hooks/use-refs';
import useScrollParent from '../hooks/use-scroll-parent';
import Sticky from '../sticky';
import TabsTitle from './TabsTitle';
import TabsContent from './TabsContent';
import TabsContext from './TabsContext';
import { addUnit, parseChildList, isHidden, unitToPx, scrollLeftTo, getElementTop, getVisibleHeight, scrollTopTo, getVisibleTop, setRootScrollTop, createNamespace, isObject } from '../utils';
import { callInterceptor } from '../utils/interceptor';
import { BORDER_TOP_BOTTOM } from '../utils/constant';
import { useUpdateEffect } from '../hooks';
import useEventListener from '../hooks/use-event-listener';
import { isReachBottom } from './utils';
import useRefState from '../hooks/use-ref-state';
const [bem] = createNamespace('tabs');

const getTabName = (tab, index) => {
  var _a;

  return (_a = tab === null || tab === void 0 ? void 0 : tab.name) !== null && _a !== void 0 ? _a : index;
};

const Tabs = forwardRef((props, ref) => {
  const {
    children,
    color,
    align,
    background
  } = props;
  const root = useRef(null);
  const swiperRef = useRef(null);
  const wrapRef = useRef(null);
  const lockScroll = useRef(false);
  const stickyFixed = useRef(false);
  const immediateRef = useRef(true);
  const navRef = useRef(null);
  const scroller = useScrollParent(root);
  const [titleRefs, setTitleRefs] = useRefs();
  const [contentRefs, setContentRefs] = useRefs();
  const wrapHeight = useMemo(() => {
    if (!wrapRef.current) return 0;
    return getVisibleHeight(wrapRef.current);
  }, [wrapRef.current]);
  const childrenList = useMemo(() => parseChildList(props.children), [props.children]);
  const defaultIndex = useMemo(() => {
    if (props.scrollspy) return 0;
    const ac = props.active === undefined ? props.defaultActive : props.active;
    let idx = childrenList.findIndex((tab, index) => getTabName(tab, index) === ac);
    if (idx < 0) idx = 0;
    return idx;
  }, [props.scrollspy, props.active, props.defaultActive]);
  const [index, setIndex] = useRefState(defaultIndex); // whether the nav is scrollable

  const scrollable = useMemo(() => childrenList.length > props.swipeThreshold || !props.ellipsis, [childrenList.length, props.swipeThreshold, props.ellipsis]);
  const navStyle = useMemo(() => ({
    borderColor: props.type === 'card' && color,
    background
  }), [color, background]);
  const currentName = useMemo(() => {
    const activeTab = childrenList === null || childrenList === void 0 ? void 0 : childrenList[index];
    return activeTab ? getTabName(activeTab, index) : 0;
  }, [childrenList, index]);
  const offsetTopPx = useMemo(() => unitToPx(props.offsetTop), [props.offsetTop]); // 下划线偏移量

  const lineTranslateLeft = useMemo(() => {
    const hidden = isHidden(root.current);
    const title = titleRefs === null || titleRefs === void 0 ? void 0 : titleRefs[index];

    if (!title || hidden || props.type !== 'line') {
      return;
    }

    return title.offsetLeft + title.offsetWidth / 2;
  }, [root.current, titleRefs, props.type, index]); // 下划线样式

  const lineStyle = useMemo(() => {
    const {
      lineWidth,
      lineHeight
    } = props;
    const measureStyle = {
      width: addUnit(lineWidth),
      backgroundColor: color,
      transitionDuration: `${immediateRef.current ? 0 : props.duration}ms`
    };

    if (lineTranslateLeft) {
      measureStyle.transform = `translateX(${lineTranslateLeft}px) translateX(-50%)`;
    }

    if (lineHeight) {
      const height = addUnit(lineHeight);
      measureStyle.height = height;
      measureStyle.borderRadius = height;
    }

    return measureStyle;
  }, [color, props.lineHeight, props.lineWidth, lineTranslateLeft, immediateRef.current]);

  const getAvailableTab = targetIndex => {
    var _a;

    const diff = targetIndex < index ? -1 : 1;

    while (targetIndex >= 0 && targetIndex < childrenList.length) {
      if (!((_a = childrenList[targetIndex]) === null || _a === void 0 ? void 0 : _a.disabled)) {
        return targetIndex;
      }

      targetIndex += diff;
    }

    return null;
  };

  const setCurrentIndex = currentIndex => {
    var _a;

    const availableIndex = getAvailableTab(currentIndex);
    if (availableIndex === null) return;
    const newTab = childrenList[availableIndex];
    const newName = getTabName(newTab, availableIndex);
    setIndex(availableIndex);
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, newName, availableIndex);
  }; // sync scrollspy content position


  const scrollToCurrentContent = current => {
    var _a, _b;

    if (!props.scrollspy) return;
    const contentTarget = contentRefs[current !== null && current !== void 0 ? current : index];

    if (contentTarget && scroller) {
      const immediate = isObject(props.scrollspy) ? (_b = (_a = props.scrollspy) === null || _a === void 0 ? void 0 : _a.scrollImmediate) !== null && _b !== void 0 ? _b : true : true; // void ui shake use math.ceil to get `to` value

      const to = Math.ceil(getElementTop(contentTarget, scroller) - (offsetTopPx + wrapHeight));
      lockScroll.current = true;
      scrollTopTo(scroller, to, immediate ? 0 : +props.duration, () => {
        lockScroll.current = false;
      });
    }
  };

  const onClickTab = (item, targetIndex, event) => {
    var _a;

    const {
      disabled = false
    } = item;
    const name = getTabName(item, targetIndex);
    (_a = props.onClickTab) === null || _a === void 0 ? void 0 : _a.call(props, {
      name,
      event,
      disabled,
      index: targetIndex
    });
    if (disabled) return;
    callInterceptor({
      interceptor: props.beforeChange,
      args: [name],
      done: () => {
        if (targetIndex !== index) {
          setCurrentIndex(targetIndex);
          scrollToCurrentContent(targetIndex);
        }
      }
    });
  }; // scroll active tab into view


  const scrollIntoView = immediate => {
    const nav = navRef.current;
    const title = titleRefs === null || titleRefs === void 0 ? void 0 : titleRefs[index];

    if (!scrollable || !nav || !title) {
      return;
    }

    const to = title.offsetLeft - (nav.offsetWidth - title.offsetWidth) / 2;
    scrollLeftTo(nav, to, immediate ? 0 : +props.duration);
  };

  const getCurrentIndexOnScroll = () => {
    const scrollOffset = offsetTopPx + wrapHeight; // eslint-disable-next-line no-plusplus

    for (let index = 0; index < contentRefs.length; index++) {
      const $el = contentRefs[index];
      const top = getVisibleTop($el);

      if (top > scrollOffset) {
        return index === 0 ? 0 : index - 1;
      }
    }

    return titleRefs.length - 1;
  };

  const onScroll = () => {
    if (!props.scrollspy || lockScroll.current) return;
    let currentIndex = getCurrentIndexOnScroll();

    if (typeof props.scrollspy === 'object') {
      if (props.scrollspy.autoFocusLast && isReachBottom(props.scrollspy.reachBottomThreshold)) {
        currentIndex = titleRefs.length - 1;
      }
    }

    if (currentIndex !== index) {
      setCurrentIndex(currentIndex);
    }
  };

  const onStickyScroll = params => {
    var _a;

    stickyFixed.current = params.isFixed;
    (_a = props.onScroll) === null || _a === void 0 ? void 0 : _a.call(props, params);
  };

  const renderNav = () => {
    return childrenList.map((item, itemIndex) => {
      return _jsx(TabsTitle, {
        ref: setTitleRefs(itemIndex),
        type: props.type,
        badge: item.badge,
        title: item.title,
        description: item.description,
        color: color,
        style: item.titleStyle,
        className: item.titleClass,
        isActive: itemIndex === index,
        disabled: item.disabled,
        scrollable: scrollable,
        activeColor: props.titleActiveColor,
        inactiveColor: props.titleInactiveColor,
        onClick: event => {
          onClickTab(item, itemIndex, event);
        }
      }, item.key);
    });
  };

  const renderHeader = () => {
    const {
      type,
      border
    } = props;
    return _jsx("div", Object.assign({
      ref: wrapRef,
      className: clsx([bem('wrap', {
        scrollable
      }), {
        [BORDER_TOP_BOTTOM]: type !== 'card' && border
      }])
    }, {
      children: _jsxs("div", Object.assign({
        ref: navRef,
        role: 'tablist',
        className: clsx(bem('nav', [type, {
          complete: scrollable,
          start: align === 'start'
        }])),
        style: navStyle
      }, {
        children: [props.navLeft, renderNav(), type === 'line' && !immediateRef.current && _jsx("div", {
          className: clsx(bem('line')),
          style: lineStyle
        }), props.navRight]
      }))
    }));
  }; // sync props.active to inner index value


  useUpdateEffect(() => {
    if (props.active === undefined) return;

    if (props.active !== currentName) {
      const currentIndex = childrenList.findIndex((tab, index) => getTabName(tab, index) === props.active);

      if (currentIndex > -1 && currentIndex !== index) {
        setIndex(currentIndex);
        scrollToCurrentContent(currentIndex);
      }
    }
  }, [props.active]); // sync tab.pane title in correct postion when index change

  useUpdateEffect(() => {
    scrollIntoView(); // scrollspy situation

    if (stickyFixed.current && props.stickyInitScrollbar && !props.scrollspy) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      setRootScrollTop(Math.ceil(getElementTop(root.current) - offsetTopPx));
    }
  }, [index]); // init

  useEffect(() => {
    immediateRef.current = false;
    scrollIntoView(true);
  }, []);
  useEventListener('scroll', onScroll, {
    target: scroller,
    depends: [index]
  });
  useImperativeHandle(ref, () => {
    var _a, _b;

    return {
      scrollTo: name => {
        const currentIndex = childrenList.findIndex((tab, index) => getTabName(tab, index) === name);

        if (currentIndex > -1 && currentIndex !== index) {
          setIndex(currentIndex);
          scrollToCurrentContent(currentIndex);
        }
      },
      swiper: swiperRef.current ? {
        enable: (_a = swiperRef.current) === null || _a === void 0 ? void 0 : _a.enable,
        disable: (_b = swiperRef.current) === null || _b === void 0 ? void 0 : _b.disable
      } : undefined
    };
  });
  return _jsx(TabsContext.Provider, Object.assign({
    value: {
      props,
      currentName,
      scrollIntoView
    }
  }, {
    children: _jsxs("div", Object.assign({
      ref: root,
      className: clsx(props.className, bem([props.type]))
    }, {
      children: [props.sticky ? _jsxs(Sticky, Object.assign({
        container: root,
        offsetTop: offsetTopPx,
        onScroll: onStickyScroll
      }, {
        children: [renderHeader(), props.navBottom]
      })) : _jsxs(_Fragment, {
        children: [renderHeader(), props.navBottom]
      }), _jsx(TabsContent, Object.assign({
        swiperRef: swiperRef,
        count: childrenList.length,
        animated: props.animated,
        duration: props.duration,
        swipeable: props.swipeable,
        lazyRender: props.lazyRender,
        currentIndex: index,
        onChange: setCurrentIndex
      }, {
        children: React.Children.toArray(children).filter(Boolean).map((node, index) => React.cloneElement(node, {
          index,
          ref: setContentRefs(index)
        }))
      }))]
    }))
  }));
});
Tabs.defaultProps = {
  type: 'line',
  duration: 300,
  swipeThreshold: 5,
  offsetTop: 0,
  ellipsis: true,
  lazyRender: true,
  stickyInitScrollbar: true,
  defaultActive: 0,
  align: 'center'
};
export default Tabs;
