// react-vant/es/tabs/TabsTitle.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useMemo, forwardRef } from 'react';
import clsx from 'clsx';
import { createNamespace, isObject } from '../utils';
import Badge from '../badge';
const [bem] = createNamespace('tab');
const TabsTitle = forwardRef((props, ref) => {
  const {
    type,
    color,
    isActive,
    activeColor,
    inactiveColor,
    disabled,
    className
  } = props;
  const customStyle = useMemo(() => {
    const style = Object.assign({}, props.style);
    const isCard = type === 'card'; // card theme color

    if (color) {
      if (isCard) {
        style.borderColor = color;
      }

      if (!disabled) {
        if (isCard) {
          if (isActive) {
            style.backgroundColor = color;
          } else {
            style.color = color;
          }
        }

        if ((props.type === 'line' || props.type === 'jumbo') && isActive) {
          style.color = color;
        }
      }
    }

    const titleColor = isActive ? activeColor : inactiveColor;

    if (titleColor) {
      style.color = titleColor;
    }

    return style;
  }, [type, color, disabled, isActive, activeColor, inactiveColor]);

  const renderText = () => {
    const Title = _jsx("span", Object.assign({
      className: clsx(bem('text', {
        ellipsis: !props.scrollable && props.type !== 'jumbo'
      })),
      style: {
        backgroundColor: props.type === 'capsule' && isActive && color
      }
    }, {
      children: (() => {
        if (typeof props.title === 'function') {
          return props.title(isActive);
        }

        return props.title;
      })()
    }));

    const Description = props.type === 'jumbo' && !!props.description ? _jsx("div", Object.assign({
      className: clsx(bem('description')),
      style: {
        backgroundColor: isActive && color
      }
    }, {
      children: (() => {
        if (typeof props.description === 'function') {
          return props.description(isActive);
        }

        return props.description;
      })()
    })) : null;

    const measureContent = _jsxs(_Fragment, {
      children: [Title, Description]
    });

    if (props.badge) {
      const badgeProps = isObject(props.badge) ? props.badge : {
        content: props.badge
      };
      return _jsx(Badge, Object.assign({}, badgeProps, {
        children: measureContent
      }));
    }

    return measureContent;
  };

  return _jsx("div", Object.assign({
    ref: ref,
    className: clsx([bem({
      active: props.isActive,
      disabled: props.disabled
    }), className]),
    style: customStyle,
    "aria-selected": props.isActive,
    onClick: props.onClick
  }, {
    children: renderText()
  }));
});
export default TabsTitle;
