// react-vant/lib/action-sheet/ActionSheet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _icons() {
  const data = require("@react-vant/icons");

  _icons = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

var _loading = _interopRequireDefault(require("../loading"));

var _popup = _interopRequireDefault(require("../popup"));

var _Popup = require("../popup/Popup");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('action-sheet');

const ActionSheet = props => {
  const onCancel = () => {
    var _a, _b;

    (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
    (_b = props.onCancel) === null || _b === void 0 ? void 0 : _b.call(props);
  };

  const renderHeader = () => {
    if (!props.title) return null;
    return (0, _jsxRuntime().jsxs)("div", Object.assign({
      className: (0, _clsx().default)(bem('header'))
    }, {
      children: [props.title, props.closeable && _react().default.cloneElement(props.closeIcon, {
        className: (0, _clsx().default)(bem('clear')),
        onClick: onCancel
      })]
    }));
  };

  const renderCancel = () => {
    if (!props.cancelText) return null;
    return [(0, _jsxRuntime().jsx)("div", {
      className: (0, _clsx().default)(bem('gap'))
    }, 'cancel-gap'), (0, _jsxRuntime().jsx)("button", Object.assign({
      type: 'button',
      className: (0, _clsx().default)(bem('cancel')),
      onClick: onCancel
    }, {
      children: props.cancelText
    }), 'cancel-btn')];
  };

  const renderOption = (item, index) => {
    const {
      name,
      color,
      subname,
      loading,
      callback,
      disabled,
      className,
      style
    } = item;
    const Content = loading ? (0, _jsxRuntime().jsx)(_loading.default, {
      className: (0, _clsx().default)(bem('loading-icon'))
    }) : [(0, _jsxRuntime().jsx)("span", Object.assign({
      className: (0, _clsx().default)(bem('name'))
    }, {
      children: name
    }), `${index}-1`), subname && (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('subname'))
    }, {
      children: subname
    }), `${index}-2`)];

    const onClick = () => {
      if (disabled || loading) {
        return;
      }

      if (callback) {
        callback(item);
      }

      if (props.closeOnClickAction) {
        onCancel();
      }

      setTimeout(() => {
        var _a;

        (_a = props.onSelect) === null || _a === void 0 ? void 0 : _a.call(props, item, index);
      }, 0);
    };

    return (0, _jsxRuntime().jsx)("button", Object.assign({
      type: 'button',
      style: Object.assign({
        color
      }, style),
      className: (0, _clsx().default)(bem('item', {
        loading,
        disabled
      }), className),
      onClick: onClick
    }, {
      children: Content
    }), index);
  };

  const renderDescription = () => {
    if (props.description) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('description'))
      }, {
        children: props.description
      }));
    }

    return null;
  };

  const renderOptions = () => {
    if (props.actions) {
      return props.actions.map(renderOption);
    }

    return null;
  };

  return (0, _jsxRuntime().jsx)(_popup.default, Object.assign({
    visible: props.visible,
    className: (0, _clsx().default)(bem('wrapper')),
    position: 'bottom'
  }, (0, _utils.pick)(props, _Popup.sharedPopupProps), {
    onClose: onCancel,
    closeable: false
  }, {
    children: (0, _jsxRuntime().jsxs)("div", Object.assign({
      className: (0, _clsx().default)(bem(), props.className),
      style: props.style
    }, {
      children: [renderHeader(), renderDescription(), (0, _jsxRuntime().jsxs)("div", Object.assign({
        className: (0, _clsx().default)(bem('content'))
      }, {
        children: [renderOptions(), props.children]
      })), renderCancel()]
    }))
  }));
};

ActionSheet.defaultProps = {
  closeable: true,
  safeAreaInsetBottom: true,
  round: true,
  lockScroll: true,
  overlay: true,
  closeOnClickOverlay: true,
  closeIcon: (0, _jsxRuntime().jsx)(_icons().Cross, {})
};
var _default = ActionSheet;
exports.default = _default;
