// react-vant/lib/sidebar/SidebarItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _badge = _interopRequireDefault(require("../badge"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('sidebar-item');

const SidebarItem = props => {
  const {
    parent,
    index
  } = props;

  const onClick = () => {
    var _a;

    if (props.disabled) {
      return;
    }

    (_a = props.onClick) === null || _a === void 0 ? void 0 : _a.call(props, index);
    parent.setActive(index);
  };

  const {
    dot,
    badge,
    title,
    disabled
  } = props;
  const selected = index === parent.getActive();
  return (0, _jsxRuntime().jsx)("div", {
    children: (0, _jsxRuntime().jsx)("a", Object.assign({
      className: (0, _clsx().default)(bem({
        select: selected,
        disabled
      })),
      onClick: onClick
    }, {
      children: (0, _jsxRuntime().jsx)(_badge.default, Object.assign({
        dot: dot,
        content: badge,
        className: (0, _clsx().default)(bem('text'))
      }, {
        children: title
      }))
    }))
  });
};

var _default = SidebarItem;
exports.default = _default;
