// react-vant/lib/sidebar/Sidebar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _useMergedState = _interopRequireDefault(require("../hooks/use-merged-state"));

var _SidebarItem = _interopRequireDefault(require("./SidebarItem"));

var _utils = require("../utils");

var _devLog = require("../utils/dev-log");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const [bem] = (0, _utils.createNamespace)('sidebar');

const Sidebar = _a => {
  var {
    children,
    className,
    style
  } = _a,
      props = (0, _tslib().__rest)(_a, ["children", "className", "style"]);
  const [active, updateActive] = (0, _useMergedState.default)({
    value: props.value,
    defaultValue: props.defaultValue
  });

  const getActive = () => active;

  const setActive = value => {
    var _a;

    if (value !== getActive()) {
      updateActive(value);
      (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, value);
    }
  };

  const validChildren = (0, _react().useMemo)(() => _react().default.Children.toArray(children).filter(Boolean).map(child => {
    if (!_react().default.isValidElement(child)) return null;

    if (child.type !== _SidebarItem.default) {
      if (process.env.NODE_ENV !== 'production') {
        (0, _devLog.devWarning)('Sidebar', ' <SidebarItem> must be a child component of <Sidebar>.');
      }

      return null;
    }

    return child;
  }), [children]);
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(className, bem('wrapper')),
    style: style
  }, {
    children: [(0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(props.sideClassName, bem()),
      style: props.sideStyle
    }, {
      children: validChildren.map((child, index) => _react().default.cloneElement(child, {
        index,
        parent: {
          setActive,
          getActive
        }
      }))
    })), validChildren.map((child, index) => {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(child.props.contentClassName, bem('content')),
        style: Object.assign(Object.assign({}, child.props.contentStyle), {
          display: index === getActive() ? undefined : 'none'
        })
      }, {
        children: child.props.children
      }), child.key);
    })]
  }));
};

Sidebar.defaultProps = {
  defaultValue: 0
};
var _default = Sidebar;
exports.default = _default;
