// react-vant/es/picker/Picker.js
import { __rest } from "tslib";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useMemo, useRef, useImperativeHandle, forwardRef, useContext, useState } from 'react';
import clsx from 'clsx';
import Loading from '../loading';
import Column from './PickerColumn';
import useEventListener from '../hooks/use-event-listener';
import { unitToPx, preventDefault, isObject, extend, createNamespace } from '../utils';
import { BORDER_UNSET_TOP_BOTTOM } from '../utils/constant';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { useMemoizedFn, usePropsValue } from '../hooks';
import Popup from '../popup';
import { useColumnsExtend } from './columnsExtend';
import useRefs from '../hooks/use-refs';
import useDebounceEffect from '../hooks/use-debunce-effect';
const [bem] = createNamespace('picker');

function PickerInner(props) {
  const {
    locale
  } = useContext(ConfigProviderContext);
  const wrapper = useRef(null);
  const [refs, setRefs] = useRefs();
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = extend({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.columnsFieldNames);
  const [innerValue, setInnerValue] = useState(props.value); // Sync `value` to `innerValue`

  useEffect(() => {
    if (props.value === undefined) return; // Uncontrolled mode

    if (JSON.stringify(innerValue) === JSON.stringify(props.value)) return;
    setInnerValue(props.value);
  }, [props.value]);
  const formatColumns = useColumnsExtend(props.columns, {
    textKey,
    valueKey,
    childrenKey
  }, innerValue);
  const {
    columns,
    items,
    indexes
  } = formatColumns;
  useDebounceEffect(() => {
    var _a;

    if (JSON.stringify(props.value) === JSON.stringify(innerValue)) return;
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, innerValue, items, indexes);
  }, [innerValue], {
    wait: 0,
    leading: false,
    trailing: true
  });
  const itemHeight = useMemo(() => unitToPx(props.itemHeight), [props.itemHeight]);

  const handleSelect = (val, index) => {
    setInnerValue(prev => {
      const next = [...prev];
      next[index] = val === null || val === void 0 ? void 0 : val[valueKey];
      return next;
    });
  };

  const confirm = () => {
    var _a;

    refs.forEach(_ref => _ref.stopMomentum());
    (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props, innerValue, items, indexes);
  };

  const cancel = () => {
    var _a;

    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(props);
  };

  const renderTitle = () => {
    if (props.title) {
      return _jsx("div", Object.assign({
        className: clsx(bem('title'), 'rv-ellipsis')
      }, {
        children: props.title
      }));
    }

    return null;
  };

  const renderCancel = () => {
    const text = props.cancelButtonText || locale.cancel;
    return _jsx("button", Object.assign({
      type: 'button',
      className: clsx(bem('cancel')),
      onClick: cancel
    }, {
      children: text
    }));
  };

  const renderConfirm = () => {
    const text = props.confirmButtonText || locale.confirm;
    return _jsx("button", Object.assign({
      type: 'button',
      className: clsx(bem('confirm')),
      onClick: confirm
    }, {
      children: text
    }));
  };

  const renderToolbar = () => {
    if (props.showToolbar) {
      return _jsx("div", Object.assign({
        className: clsx(bem('toolbar'))
      }, {
        children: props.toolbar || _jsxs(_Fragment, {
          children: [renderCancel(), renderTitle(), renderConfirm()]
        })
      }));
    }

    return null;
  };

  const renderColumnItems = () => {
    return columns.map((item, columnIndex) => {
      const placeholder = Array.isArray(props.placeholder) ? props.placeholder[columnIndex] : props.placeholder;
      return _jsx(Column, {
        textKey: textKey,
        valueKey: valueKey,
        ref: setRefs(columnIndex),
        placeholder: placeholder,
        optionRender: props.optionRender,
        readOnly: props.readOnly,
        value: innerValue[columnIndex],
        itemHeight: itemHeight,
        index: columnIndex,
        swipeDuration: props.swipeDuration,
        visibleItemCount: props.visibleItemCount,
        options: item,
        onSelect: handleSelect
      }, columnIndex);
    });
  };

  const renderColumns = () => {
    const wrapHeight = itemHeight * props.visibleItemCount;
    const frameStyle = {
      height: `${itemHeight}px`
    };
    const columnsStyle = {
      height: `${wrapHeight}px`
    };
    const maskStyle = {
      backgroundSize: `100% ${(wrapHeight - itemHeight) / 2}px`
    };
    return _jsxs("div", Object.assign({
      ref: wrapper,
      className: clsx(bem('columns')),
      style: columnsStyle
    }, {
      children: [renderColumnItems(), _jsx("div", {
        className: clsx(bem('mask')),
        style: maskStyle
      }), _jsx("div", {
        className: clsx(BORDER_UNSET_TOP_BOTTOM, bem('frame')),
        style: frameStyle
      })]
    }));
  };

  useEventListener('touchmove', preventDefault, {
    target: wrapper.current
  });
  return _jsxs("div", Object.assign({
    className: clsx(bem(), props.className)
  }, {
    children: [props.toolbarPosition === 'top' ? renderToolbar() : null, props.loading ? _jsx(Loading, {
      className: clsx(bem('loading'))
    }) : null, props.columnsTop, renderColumns(), props.columnsBottom, props.toolbarPosition === 'bottom' ? renderToolbar() : null]
  }));
}

function PopupPicker(props, ref) {
  const {
    visible: outerVisible,
    popup,
    children,
    defaultValue = []
  } = props,
        pickerProps = __rest(props, ["visible", "popup", "children", "defaultValue"]);

  const [visible, setVisible] = usePropsValue({
    value: outerVisible,
    defaultValue: false,
    onChange: v => {
      var _a;

      if (v === false) {
        (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
      }
    }
  });
  const actions = {
    toggle: () => {
      if (popup) setVisible(v => !v);
    },
    open: () => {
      if (popup) {
        setVisible(true);
      }
    },
    close: () => {
      if (popup) {
        setVisible(false);
      }
    }
  };
  useImperativeHandle(ref, () => actions);
  const formatValue = Array.isArray(props.value) ? props.value : props.value !== undefined ? [props.value] : undefined;
  const formatDefaultValue = Array.isArray(defaultValue) ? defaultValue : defaultValue !== undefined ? [defaultValue] : [];
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = extend({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.columnsFieldNames);
  const isPlainType = useMemo(() => {
    const firstColumn = props.columns[0] || {};
    if (Array.isArray(firstColumn)) return false;

    if (typeof firstColumn === 'object') {
      // 联级
      if (childrenKey in firstColumn) {
        return false;
      }
    } // 单列


    return true;
  }, [props.columns, childrenKey]);

  const parseValue = target => isPlainType ? target === null || target === void 0 ? void 0 : target[0] : target;

  const [value, setValue] = usePropsValue({
    value: formatValue,
    defaultValue: formatDefaultValue
  });
  const formatColumns = useColumnsExtend(props.columns, {
    textKey,
    valueKey,
    childrenKey
  }, value);
  const [innerValue, setInnerValue] = useState(value);
  useEffect(() => {
    if (popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [visible]);
  useEffect(() => {
    if (!popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [value]);

  const onConfirm = (val, items, indexes) => {
    var _a;

    setValue(innerValue, true);
    (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props, parseValue(val), parseValue(items), parseValue(indexes));
    if (popup) actions.close();
  };

  const onCancel = () => {
    var _a;

    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(props);
    if (popup) actions.close();
  };

  const onChange = useMemoizedFn((val, ext, indexes) => {
    var _a, _b;

    setInnerValue(val);

    if (popup) {
      if (visible) (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, parseValue(val), parseValue(ext), parseValue(indexes));
    } else {
      (_b = props.onChange) === null || _b === void 0 ? void 0 : _b.call(props, parseValue(val), parseValue(ext), parseValue(indexes));
    }
  });
  const popupProps = isObject(popup) ? Object.assign({
    closeOnClickOverlay: true
  }, popup) : {
    closeOnClickOverlay: true
  };

  const content = _jsx(PickerInner, Object.assign({}, pickerProps, {
    value: innerValue,
    onCancel: onCancel,
    onConfirm: onConfirm,
    onChange: onChange
  }));

  if (!popup) return content;
  return _jsxs(_Fragment, {
    children: [_jsx(Popup, Object.assign({
      round: true,
      position: 'bottom',
      visible: visible,
      closeOnClickOverlay: true,
      onClickOverlay: () => {
        if (!(popupProps === null || popupProps === void 0 ? void 0 : popupProps.closeOnClickOverlay)) return;
        setVisible(false);
      }
    }, popupProps, {
      children: content
    })), children === null || children === void 0 ? void 0 : children(parseValue(value), parseValue(formatColumns.items), actions)]
  });
}

const Picker = forwardRef(PopupPicker);
Picker.defaultProps = {
  columns: [],
  itemHeight: 44,
  visibleItemCount: 5,
  swipeDuration: 300,
  showToolbar: true,
  placeholder: true,
  toolbarPosition: 'top'
};
export default Picker;
