// react-vant/lib/uploader/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filterFiles = filterFiles;
exports.isImageFile = isImageFile;
exports.isImageUrl = isImageUrl;
exports.isOversize = isOversize;
exports.readFileContent = readFileContent;
exports.toArray = toArray;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function toArray(item) {
  if (Array.isArray(item)) {
    return item;
  }

  return [item];
}

function readFileContent(file, resultType) {
  return new Promise(resolve => {
    if (resultType === 'file') {
      resolve();
      return;
    }

    if (isImageFile(file)) {
      resolve(URL.createObjectURL(file));
      return;
    }

    const reader = new FileReader();

    reader.onload = event => {
      resolve(event.target.result);
    };

    if (resultType === 'dataUrl') {
      reader.readAsDataURL(file);
    } else if (resultType === 'text') {
      reader.readAsText(file);
    }
  });
}

function isOversize(files, maxSize) {
  return toArray(files).some(file => {
    if (file) {
      if ((0, _utils.isFunction)(maxSize)) {
        return maxSize(file);
      }

      return file.size > maxSize;
    }

    return false;
  });
}

function filterFiles(items, maxSize) {
  const valid = [];
  const invalid = [];
  items.forEach(item => {
    if (isOversize([item], maxSize)) {
      invalid.push(item);
    } else {
      valid.push(item);
    }
  });
  return {
    valid,
    invalid
  };
}

const IMAGE_REGEXP = /\.(jpeg|jpg|gif|png|svg|webp|jfif|bmp|dpg)/i;

function isImageUrl(url) {
  return IMAGE_REGEXP.test(url);
}

function isImageFile(item, isImage) {
  // some special urls cannot be recognized
  // user can add `isImage` flag to mark it as an image url
  if (isImage) {
    return true;
  }

  if (item.file && item.file.type) {
    return item.file.type.indexOf('image') === 0;
  }

  if (item.url) {
    return isImageUrl(item.url);
  }

  if (item.thumbnail) {
    return isImageUrl(item.thumbnail);
  }

  return false;
}
