// react-vant/lib/calendar/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calcDateNum = calcDateNum;
exports.cloneDates = exports.cloneDate = void 0;
exports.compareDay = compareDay;
exports.compareMonth = compareMonth;
exports.formatMonthTitle = formatMonthTitle;
exports.getDayByOffset = getDayByOffset;
exports.getToday = exports.getPrevDay = exports.getNextDay = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-nested-ternary */
function formatMonthTitle(date) {
  return [date.getFullYear(), date.getMonth() + 1];
}

function compareMonth(date1, date2) {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

function compareDay(day1, day2) {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

const cloneDate = date => new Date(date);

exports.cloneDate = cloneDate;

const cloneDates = dates => Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

exports.cloneDates = cloneDates;

function getDayByOffset(date, offset) {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

const getPrevDay = date => getDayByOffset(date, -1);

exports.getPrevDay = getPrevDay;

const getNextDay = date => getDayByOffset(date, 1);

exports.getNextDay = getNextDay;

const getToday = () => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

exports.getToday = getToday;

function calcDateNum(date) {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}
