// react-vant/es/input/Input.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React, { useState, forwardRef, useImperativeHandle, useMemo, useRef } from 'react';
import { Clear } from '@react-vant/icons';
import clsx from 'clsx';
import { isDef, formatNumber, preventDefault, resetScroll, createNamespace } from '../utils';
import { usePropsValue } from '../hooks';
const [bem] = createNamespace('input');
const Input = forwardRef((props, ref) => {
  const inputRef = useRef();
  const [inputFocus, setInputFocus] = useState(false);
  const compositionStartRef = useRef(false);
  const [value, setValue] = usePropsValue(props);
  const {
    className,
    style,
    align,
    type,
    name,
    placeholder,
    disabled,
    readOnly,
    maxLength,
    autoFocus
  } = props;

  const focus = () => {
    if (inputRef === null || inputRef === void 0 ? void 0 : inputRef.current) {
      inputRef.current.focus();
    }
  };

  const blur = () => {
    if (inputRef === null || inputRef === void 0 ? void 0 : inputRef.current) {
      inputRef.current.blur();
    }
  };

  useImperativeHandle(ref, () => ({
    clear: () => {
      setValue('');
    },
    focus,
    blur,

    get nativeElement() {
      return inputRef.current;
    }

  }));
  const showClear = useMemo(() => {
    if (props.clearable && !readOnly) {
      const hasValue = value !== '';
      const trigger = props.clearTrigger === 'always' || props.clearTrigger === 'focus' && inputFocus;
      return hasValue && trigger;
    }

    return false;
  }, [value, props.clearTrigger, inputFocus]);

  const handleChange = e => {
    var _a, _b;

    const inputValue = (_a = e === null || e === void 0 ? void 0 : e.currentTarget) === null || _a === void 0 ? void 0 : _a.value;
    let finalValue = inputValue;

    if (isDef(maxLength) && finalValue.length > +maxLength) {
      finalValue = finalValue.slice(0, maxLength);
      (_b = props.onOverlimit) === null || _b === void 0 ? void 0 : _b.call(props);
    }

    if (type === 'number' || type === 'digit') {
      const isNumber = type === 'number';
      finalValue = formatNumber(finalValue, isNumber, isNumber);
    }

    setValue(finalValue);
  };

  const handleFocus = e => {
    var _a;

    setInputFocus(true);
    (_a = props.onFocus) === null || _a === void 0 ? void 0 : _a.call(props, e); // readOnly not work in legacy mobile safari

    if (readOnly) {
      blur();
    }
  };

  const handleBulr = e => {
    var _a;

    setInputFocus(false);
    (_a = props.onBlur) === null || _a === void 0 ? void 0 : _a.call(props, e);
    resetScroll();
  };

  const handleKeyPress = e => {
    var _a;

    if (e.key === 'Enter' || +e.charCode === 13) {
      preventDefault(e); // trigger blur after click keyboard search button

      if (props.type === 'search') {
        blur();
      }
    }

    (_a = props.onKeyPress) === null || _a === void 0 ? void 0 : _a.call(props, e);
  };

  const renderInput = () => {
    let inputType = type;
    let inputMode; // type="number" is weired in iOS, and can't prevent dot in Android
    // so use inputmode to set keyboard in mordern browers

    if (type === 'number') {
      inputType = 'text';
      inputMode = 'decimal';
    }

    if (type === 'digit') {
      inputType = 'tel';
      inputMode = 'numeric';
    }

    return _jsx("input", {
      value: value,
      type: inputType,
      inputMode: inputMode,
      ref: inputRef,
      name: name,
      className: clsx(bem('control')),
      disabled: disabled,
      autoFocus: autoFocus,
      readOnly: readOnly,
      placeholder: placeholder || '',
      onBlur: handleBulr,
      onFocus: handleFocus,
      onChange: handleChange,
      onKeyPress: handleKeyPress,
      autoCapitalize: props.autoCapitalize,
      autoCorrect: props.autoCorrect,
      onKeyDown: props.onKeyDown,
      onKeyUp: props.onKeyUp,
      onCompositionStart: e => {
        var _a;

        compositionStartRef.current = true;
        (_a = props.onCompositionStart) === null || _a === void 0 ? void 0 : _a.call(props, e);
      },
      onCompositionEnd: e => {
        var _a;

        compositionStartRef.current = false;
        (_a = props.onCompositionEnd) === null || _a === void 0 ? void 0 : _a.call(props, e);
      },
      onClick: props.onClick
    });
  };

  const handleClear = e => {
    var _a;

    setValue('');
    (_a = props.onClear) === null || _a === void 0 ? void 0 : _a.call(props, e);
  };

  return _jsxs("div", Object.assign({
    className: clsx(bem([align]), className),
    style: style
  }, {
    children: [props.prefix && _jsx("div", Object.assign({
      className: clsx(bem('prefix'))
    }, {
      children: props.prefix
    })), renderInput(), showClear && React.cloneElement(props.clearIcon, {
      className: clsx(bem('clear')),
      onTouchStart: handleClear
    }), props.suffix && _jsx("div", Object.assign({
      className: clsx(bem('suffix'))
    }, {
      children: props.suffix
    }))]
  }));
});
Input.defaultProps = {
  clearIcon: _jsx(Clear, {}),
  clearTrigger: 'focus',
  defaultValue: ''
};
export default Input;
