// react-vant/lib/steps/Steps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

function _icons() {
  const data = require("@react-vant/icons");

  _icons = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('steps');

const Steps = _a => {
  var {
    children,
    className,
    style
  } = _a,
      props = (0, _tslib().__rest)(_a, ["children", "className", "style"]);
  return (0, _jsxRuntime().jsx)("div", Object.assign({
    className: (0, _clsx().default)(className, bem([props.direction])),
    style: style
  }, {
    children: (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('items'))
    }, {
      children: _react().default.Children.toArray(children).filter(Boolean).map((child, index) => _react().default.cloneElement(child, {
        index,
        parent: props
      }))
    }))
  }));
};

Steps.defaultProps = {
  active: 0,
  direction: 'horizontal',
  activeIcon: (0, _jsxRuntime().jsx)(_icons().Checked, {})
};
var _default = Steps;
exports.default = _default;
