// react-vant/lib/picker/Picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _loading = _interopRequireDefault(require("../loading"));

var _PickerColumn = _interopRequireDefault(require("./PickerColumn"));

var _useEventListener = _interopRequireDefault(require("../hooks/use-event-listener"));

var _utils = require("../utils");

var _constant = require("../utils/constant");

var _ConfigProviderContext = _interopRequireDefault(require("../config-provider/ConfigProviderContext"));

var _hooks = require("../hooks");

var _popup = _interopRequireDefault(require("../popup"));

var _columnsExtend = require("./columnsExtend");

var _useRefs = _interopRequireDefault(require("../hooks/use-refs"));

var _useDebunceEffect = _interopRequireDefault(require("../hooks/use-debunce-effect"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const [bem] = (0, _utils.createNamespace)('picker');

function PickerInner(props) {
  const {
    locale
  } = (0, _react().useContext)(_ConfigProviderContext.default);
  const wrapper = (0, _react().useRef)(null);
  const [refs, setRefs] = (0, _useRefs.default)();
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = (0, _utils.extend)({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.columnsFieldNames);
  const [innerValue, setInnerValue] = (0, _react().useState)(props.value); // Sync `value` to `innerValue`

  (0, _react().useEffect)(() => {
    if (props.value === undefined) return; // Uncontrolled mode

    if (JSON.stringify(innerValue) === JSON.stringify(props.value)) return;
    setInnerValue(props.value);
  }, [props.value]);
  const formatColumns = (0, _columnsExtend.useColumnsExtend)(props.columns, {
    textKey,
    valueKey,
    childrenKey
  }, innerValue);
  const {
    columns,
    items,
    indexes
  } = formatColumns;
  (0, _useDebunceEffect.default)(() => {
    var _a;

    if (JSON.stringify(props.value) === JSON.stringify(innerValue)) return;
    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, innerValue, items, indexes);
  }, [innerValue], {
    wait: 0,
    leading: false,
    trailing: true
  });
  const itemHeight = (0, _react().useMemo)(() => (0, _utils.unitToPx)(props.itemHeight), [props.itemHeight]);

  const handleSelect = (val, index) => {
    setInnerValue(prev => {
      const next = [...prev];
      next[index] = val === null || val === void 0 ? void 0 : val[valueKey];
      return next;
    });
  };

  const confirm = () => {
    var _a;

    refs.forEach(_ref => _ref.stopMomentum());
    (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props, innerValue, items, indexes);
  };

  const cancel = () => {
    var _a;

    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(props);
  };

  const renderTitle = () => {
    if (props.title) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('title'), 'rv-ellipsis')
      }, {
        children: props.title
      }));
    }

    return null;
  };

  const renderCancel = () => {
    const text = props.cancelButtonText || locale.cancel;
    return (0, _jsxRuntime().jsx)("button", Object.assign({
      type: 'button',
      className: (0, _clsx().default)(bem('cancel')),
      onClick: cancel
    }, {
      children: text
    }));
  };

  const renderConfirm = () => {
    const text = props.confirmButtonText || locale.confirm;
    return (0, _jsxRuntime().jsx)("button", Object.assign({
      type: 'button',
      className: (0, _clsx().default)(bem('confirm')),
      onClick: confirm
    }, {
      children: text
    }));
  };

  const renderToolbar = () => {
    if (props.showToolbar) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('toolbar'))
      }, {
        children: props.toolbar || (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
          children: [renderCancel(), renderTitle(), renderConfirm()]
        })
      }));
    }

    return null;
  };

  const renderColumnItems = () => {
    return columns.map((item, columnIndex) => {
      const placeholder = Array.isArray(props.placeholder) ? props.placeholder[columnIndex] : props.placeholder;
      return (0, _jsxRuntime().jsx)(_PickerColumn.default, {
        textKey: textKey,
        valueKey: valueKey,
        ref: setRefs(columnIndex),
        placeholder: placeholder,
        optionRender: props.optionRender,
        readOnly: props.readOnly,
        value: innerValue[columnIndex],
        itemHeight: itemHeight,
        index: columnIndex,
        swipeDuration: props.swipeDuration,
        visibleItemCount: props.visibleItemCount,
        options: item,
        onSelect: handleSelect
      }, columnIndex);
    });
  };

  const renderColumns = () => {
    const wrapHeight = itemHeight * props.visibleItemCount;
    const frameStyle = {
      height: `${itemHeight}px`
    };
    const columnsStyle = {
      height: `${wrapHeight}px`
    };
    const maskStyle = {
      backgroundSize: `100% ${(wrapHeight - itemHeight) / 2}px`
    };
    return (0, _jsxRuntime().jsxs)("div", Object.assign({
      ref: wrapper,
      className: (0, _clsx().default)(bem('columns')),
      style: columnsStyle
    }, {
      children: [renderColumnItems(), (0, _jsxRuntime().jsx)("div", {
        className: (0, _clsx().default)(bem('mask')),
        style: maskStyle
      }), (0, _jsxRuntime().jsx)("div", {
        className: (0, _clsx().default)(_constant.BORDER_UNSET_TOP_BOTTOM, bem('frame')),
        style: frameStyle
      })]
    }));
  };

  (0, _useEventListener.default)('touchmove', _utils.preventDefault, {
    target: wrapper.current
  });
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(bem(), props.className)
  }, {
    children: [props.toolbarPosition === 'top' ? renderToolbar() : null, props.loading ? (0, _jsxRuntime().jsx)(_loading.default, {
      className: (0, _clsx().default)(bem('loading'))
    }) : null, props.columnsTop, renderColumns(), props.columnsBottom, props.toolbarPosition === 'bottom' ? renderToolbar() : null]
  }));
}

function PopupPicker(props, ref) {
  const {
    visible: outerVisible,
    popup,
    children,
    defaultValue = []
  } = props,
        pickerProps = (0, _tslib().__rest)(props, ["visible", "popup", "children", "defaultValue"]);
  const [visible, setVisible] = (0, _hooks.usePropsValue)({
    value: outerVisible,
    defaultValue: false,
    onChange: v => {
      var _a;

      if (v === false) {
        (_a = props.onClose) === null || _a === void 0 ? void 0 : _a.call(props);
      }
    }
  });
  const actions = {
    toggle: () => {
      if (popup) setVisible(v => !v);
    },
    open: () => {
      if (popup) {
        setVisible(true);
      }
    },
    close: () => {
      if (popup) {
        setVisible(false);
      }
    }
  };
  (0, _react().useImperativeHandle)(ref, () => actions);
  const formatValue = Array.isArray(props.value) ? props.value : props.value !== undefined ? [props.value] : undefined;
  const formatDefaultValue = Array.isArray(defaultValue) ? defaultValue : defaultValue !== undefined ? [defaultValue] : [];
  const {
    text: textKey,
    value: valueKey,
    children: childrenKey
  } = (0, _utils.extend)({
    text: 'text',
    value: 'value',
    children: 'children'
  }, props.columnsFieldNames);
  const isPlainType = (0, _react().useMemo)(() => {
    const firstColumn = props.columns[0] || {};
    if (Array.isArray(firstColumn)) return false;

    if (typeof firstColumn === 'object') {
      // 联级
      if (childrenKey in firstColumn) {
        return false;
      }
    } // 单列


    return true;
  }, [props.columns, childrenKey]);

  const parseValue = target => isPlainType ? target === null || target === void 0 ? void 0 : target[0] : target;

  const [value, setValue] = (0, _hooks.usePropsValue)({
    value: formatValue,
    defaultValue: formatDefaultValue
  });
  const formatColumns = (0, _columnsExtend.useColumnsExtend)(props.columns, {
    textKey,
    valueKey,
    childrenKey
  }, value);
  const [innerValue, setInnerValue] = (0, _react().useState)(value);
  (0, _react().useEffect)(() => {
    if (popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [visible]);
  (0, _react().useEffect)(() => {
    if (!popup && JSON.stringify(innerValue) !== JSON.stringify(value)) {
      setInnerValue(value);
    }
  }, [value]);

  const onConfirm = (val, items, indexes) => {
    var _a;

    setValue(innerValue, true);
    (_a = props.onConfirm) === null || _a === void 0 ? void 0 : _a.call(props, parseValue(val), parseValue(items), parseValue(indexes));
    if (popup) actions.close();
  };

  const onCancel = () => {
    var _a;

    (_a = props.onCancel) === null || _a === void 0 ? void 0 : _a.call(props);
    if (popup) actions.close();
  };

  const onChange = (0, _hooks.useMemoizedFn)((val, ext, indexes) => {
    var _a, _b;

    setInnerValue(val);

    if (popup) {
      if (visible) (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, parseValue(val), parseValue(ext), parseValue(indexes));
    } else {
      (_b = props.onChange) === null || _b === void 0 ? void 0 : _b.call(props, parseValue(val), parseValue(ext), parseValue(indexes));
    }
  });
  const popupProps = (0, _utils.isObject)(popup) ? Object.assign({
    closeOnClickOverlay: true
  }, popup) : {
    closeOnClickOverlay: true
  };
  const content = (0, _jsxRuntime().jsx)(PickerInner, Object.assign({}, pickerProps, {
    value: innerValue,
    onCancel: onCancel,
    onConfirm: onConfirm,
    onChange: onChange
  }));
  if (!popup) return content;
  return (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
    children: [(0, _jsxRuntime().jsx)(_popup.default, Object.assign({
      round: true,
      position: 'bottom',
      visible: visible,
      closeOnClickOverlay: true,
      onClickOverlay: () => {
        if (!(popupProps === null || popupProps === void 0 ? void 0 : popupProps.closeOnClickOverlay)) return;
        setVisible(false);
      }
    }, popupProps, {
      children: content
    })), children === null || children === void 0 ? void 0 : children(parseValue(value), parseValue(formatColumns.items), actions)]
  });
}

const Picker = (0, _react().forwardRef)(PopupPicker);
Picker.defaultProps = {
  columns: [],
  itemHeight: 44,
  visibleItemCount: 5,
  swipeDuration: 300,
  showToolbar: true,
  placeholder: true,
  toolbarPosition: 'top'
};
var _default = Picker;
exports.default = _default;
