// react-vant/lib/flex/FlexItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _FlexContext = _interopRequireDefault(require("./FlexContext"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('flexitem');

const FlexItem = props => {
  const {
    style,
    className,
    span,
    children,
    flex
  } = props,
        others = (0, _tslib().__rest)(props, ["style", "className", "span", "children", "flex"]);
  const classes = (0, _clsx().default)(bem([span === null || span === void 0 ? void 0 : span.toString()]), className);

  const parseFlex = _flex => {
    if (typeof _flex === 'number') {
      return `${_flex} ${_flex} auto`;
    }

    if (/^\d+(\.\d+)?(px|em|rem|%)$/.test(_flex)) {
      return `0 0 ${_flex}`;
    }

    return _flex;
  };

  return (0, _jsxRuntime().jsx)(_FlexContext.default.Consumer, {
    children: ({
      gutter
    }) => {
      let mergedStyle = Object.assign({}, style);

      if (gutter) {
        mergedStyle = Object.assign(Object.assign(Object.assign({}, gutter[0] > 0 ? {
          paddingLeft: gutter[0] / 2,
          paddingRight: gutter[0] / 2
        } : {}), gutter[1] > 0 ? {
          paddingTop: gutter[1] / 2,
          paddingBottom: gutter[1] / 2
        } : {}), mergedStyle);
      }

      if (flex) {
        mergedStyle.flex = parseFlex(flex);
      }

      return (0, _jsxRuntime().jsx)("div", Object.assign({}, others, {
        style: mergedStyle,
        className: classes
      }, {
        children: children
      }));
    }
  });
};

var _default = FlexItem;
exports.default = _default;
