// react-vant/es/selector/Selector.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import clsx from 'clsx';
import { usePropsValue } from '../hooks';
import { CheckMark } from './CheckMark';
import { createNamespace } from '../utils'; // https://mobile.ant.design/zh/components/selector

const [bem] = createNamespace('selector');
const defaultProps = {
  multiple: false,
  defaultValue: [],
  showCheckMark: true
};
export const Selector = p => {
  const props = Object.assign(Object.assign({}, defaultProps), p);
  const [value, setValue] = usePropsValue({
    value: props.value,
    defaultValue: props.defaultValue,
    onChange: val => {
      var _a;

      const extend = {
        get items() {
          return props.options.filter(option => val.includes(option.value));
        }

      };
      (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, val, extend);
    }
  });
  const items = props.options.map(option => {
    const active = (value || []).includes(option.value);
    const disabled = option.disabled || props.disabled;
    const itemCls = clsx(bem('item', {
      active: active && !props.multiple,
      'multiple-active': active && props.multiple,
      disabled
    }));
    return _jsxs("div", Object.assign({
      className: itemCls,
      onClick: () => {
        if (disabled) {
          return;
        }

        if (props.multiple) {
          const val = active ? value.filter(v => v !== option.value) : [...value, option.value];
          setValue(val);
        } else {
          const val = active ? [] : [option.value];
          setValue(val);
        }
      }
    }, {
      children: [option.label, option.description && _jsx("div", Object.assign({
        className: clsx(bem('item-description'))
      }, {
        children: option.description
      })), active && props.showCheckMark && _jsx("div", Object.assign({
        className: clsx(bem('check-mark-wrapper'))
      }, {
        children: _jsx(CheckMark, {})
      }))]
    }), option.value);
  });
  return _jsx("div", Object.assign({
    className: clsx(bem(), props.className),
    style: props.style
  }, {
    children: items
  }));
};
