// react-vant/es/field/Field.js
import { __rest } from "tslib";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import { Clear, QuestionO } from '@react-vant/icons';
import clsx from 'clsx';
import Cell from '../cell';
import Dialog from '../dialog';
import Input from '../input';
import { isDef, addUnit, createNamespace } from '../utils';
import { COMPONENT_TYPE_KEY } from '../utils/constant';
const [bem] = createNamespace('field');
const Field = forwardRef((props, ref) => {
  var _a;

  const inputRef = useRef(null);
  const textareaRef = useRef(null);
  const elementRef = props.type === 'textarea' ? textareaRef : inputRef;

  const focus = () => {
    if (elementRef.current) {
      elementRef.current.focus();
    }
  };

  const blur = () => {
    if (elementRef.current) {
      elementRef.current.blur();
    }
  };

  const clear = () => {
    if (elementRef.current) {
      elementRef.current.clear();
    }
  };

  useImperativeHandle(ref, () => ({
    focus,
    blur,
    clear,

    get nativeElement() {
      return elementRef.current.nativeElement;
    }

  }));

  const getProp = key => {
    if (isDef(props[key])) {
      return props[key];
    }

    return null;
  };

  const labelStyle = () => {
    const labelW = getProp('labelWidth');

    if (labelW) {
      return {
        width: addUnit(labelW)
      };
    }

    return {};
  };

  const formatValue = (inputValue, trigger = 'onChange') => {
    const {
      formatTrigger,
      formatter
    } = props;

    if (formatter && trigger === formatTrigger) {
      return formatter(inputValue);
    }

    return inputValue;
  };

  const onChange = val => {
    var _a;

    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, formatValue(val));
  };

  const renderInput = () => {
    const {
      value,
      defaultValue,
      align,
      type,
      placeholder,
      name,
      maxLength,
      disabled,
      readOnly,
      clearable,
      clearIcon,
      clearTrigger,
      autoFocus,
      onClear,
      onBlur,
      onFocus,
      onKeyPress,
      onOverlimit
    } = props;

    if (props.children) {
      return _jsx("div", Object.assign({
        className: clsx(bem('children'))
      }, {
        children: props.children
      }));
    }

    const commonProps = {
      value,
      onChange,
      placeholder,
      name,
      defaultValue,
      disabled,
      clearable,
      clearIcon,
      clearTrigger,
      onClear,
      onBlur,
      onFocus,
      onKeyPress,
      onOverlimit,
      autoFocus,
      readOnly,
      maxLength,
      onClick: props.onClickInput
    };

    if (type === 'textarea') {
      return _jsx(Input.TextArea, Object.assign({
        ref: textareaRef,
        autoSize: props.autoSize,
        showWordLimit: props.showWordLimit,
        rows: props.rows
      }, commonProps));
    }

    return _jsx(Input, Object.assign({
      ref: inputRef,
      type: type,
      align: align
    }, commonProps));
  };

  const renderLeftIcon = () => {
    const {
      leftIcon,
      onClickLeftIcon
    } = props;
    if (!leftIcon) return null;
    return _jsx("div", Object.assign({
      className: clsx(bem('left-icon')),
      onClick: onClickLeftIcon
    }, {
      children: leftIcon
    }));
  };

  const renderRightIcon = () => {
    const {
      rightIcon,
      onClickRightIcon
    } = props;
    if (!rightIcon) return null;
    return _jsx("div", Object.assign({
      className: clsx(bem('right-icon')),
      onClick: onClickRightIcon
    }, {
      children: rightIcon
    }));
  };

  const renderMessage = () => {
    const message = props.errorMessage;

    if (message) {
      return _jsx("div", Object.assign({
        className: clsx(bem('error-message'))
      }, {
        children: message
      }));
    }

    return null;
  };

  const renderIntro = () => {
    if (props.intro) {
      return _jsx("div", Object.assign({
        className: clsx(bem('intro'))
      }, {
        children: props.intro
      }));
    }

    return null;
  };

  const renderTooltip = () => {
    const {
      tooltip
    } = props;

    if (tooltip) {
      let icon = _jsx(QuestionO, {});

      let dialogProps = {
        message: tooltip
      };

      if (!(React.isValidElement(tooltip) || typeof tooltip === 'string')) {
        const _a = tooltip,
              {
          icon: customIcon
        } = _a,
              customDialogProps = __rest(_a, ["icon"]);

        icon = customIcon || icon;
        dialogProps = customDialogProps;
      }

      return _jsx("div", Object.assign({
        className: clsx(bem('tooltip')),
        onClick: () => Dialog.show(dialogProps)
      }, {
        children: icon
      }));
    }

    return null;
  };

  const renderLabel = () => {
    const {
      label,
      colon
    } = props;

    if (label) {
      return _jsxs(_Fragment, {
        children: [label, !!colon && ':', renderTooltip()]
      });
    }

    return null;
  };

  const {
    size,
    center,
    border,
    isLink,
    required,
    clickable,
    labelAlign,
    className,
    labelClass,
    valueClass,
    controlAlign,
    arrowDirection,
    disabled,
    titleStyle,
    error
  } = props;
  const suffix = (_a = props.suffix) !== null && _a !== void 0 ? _a : props.button;
  return _jsxs(Cell, Object.assign({
    title: renderLabel(),
    size: size,
    icon: renderLeftIcon(),
    center: center,
    border: border,
    isLink: isLink,
    required: required,
    clickable: clickable,
    extra: props.extra,
    titleStyle: Object.assign(Object.assign({}, labelStyle()), titleStyle),
    valueClass: clsx(bem('value', [controlAlign]), valueClass),
    titleClass: clsx(bem('label', labelAlign), labelClass),
    arrowDirection: arrowDirection,
    onClick: props.onClick,
    style: props.style,
    className: clsx(bem({
      error,
      disabled,
      [`label-${labelAlign}`]: labelAlign
    }), className)
  }, {
    children: [_jsxs("div", Object.assign({
      className: clsx(bem('body'))
    }, {
      children: [props.prefix && _jsx("div", Object.assign({
        className: clsx(bem('prefix'))
      }, {
        children: props.prefix
      })), _jsx("div", Object.assign({
        className: clsx(bem('control-wrapper'))
      }, {
        children: renderInput()
      })), renderRightIcon(), suffix && _jsx("div", Object.assign({
        className: clsx(bem('suffix'))
      }, {
        children: suffix
      }))]
    })), renderMessage(), renderIntro()]
  }));
});
Field.defaultProps = {
  clearIcon: _jsx(Clear, {}),
  clearTrigger: 'focus',
  formatTrigger: 'onChange',
  defaultValue: ''
};
export const FIELD_KEY = Symbol('field');
const FieldNamespace = Object.assign(Field, {
  [COMPONENT_TYPE_KEY]: FIELD_KEY
});
export default FieldNamespace;
