// react-vant/lib/list/LoadMore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _utils = require("../utils");

var _useLockFn = _interopRequireDefault(require("../hooks/use-lock-fn"));

var _useScrollParent = require("../hooks/use-scroll-parent");

var _loading = _interopRequireDefault(require("../loading"));

var _ConfigProviderContext = _interopRequireDefault(require("../config-provider/ConfigProviderContext"));

var _useThrottleFn = _interopRequireDefault(require("../hooks/use-throttle-fn"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const [bem] = (0, _utils.createNamespace)('list');

function isWindow(element) {
  return element === window;
}

const LoadMore = _react().default.forwardRef((props, ref) => {
  const {
    locale
  } = _react().default.useContext(_ConfigProviderContext.default);

  const [failed, setFailed] = _react().default.useState(false);

  const doLoadMore = (0, _useLockFn.default)(isRetry => (0, _tslib().__awaiter)(void 0, void 0, void 0, function* () {
    try {
      yield props.onLoad(isRetry);
    } catch (e) {
      setFailed(true);
      throw e;
    }
  }));

  const elementRef = _react().default.useRef(null); // Prevent duplicated trigger of `check` function


  const [flag, setFlag] = _react().default.useState({});

  const nextFlagRef = _react().default.useRef(flag);

  const [scrollParent, setScrollParent] = _react().default.useState();

  const {
    run: check
  } = (0, _useThrottleFn.default)(() => (0, _tslib().__awaiter)(void 0, void 0, void 0, function* () {
    if (nextFlagRef.current !== flag) return;
    if (props.finished) return;
    const element = elementRef.current;
    if (!element) return;
    if (!element.offsetParent) return;
    const parent = (0, _useScrollParent.getScrollParent)(element);
    setScrollParent(parent);
    if (!parent) return;
    const rect = element.getBoundingClientRect();
    const elementTop = rect.top;
    const current = isWindow(parent) ? window.innerHeight : parent.getBoundingClientRect().bottom;
    const isReachEdge = current >= elementTop - props.threshold;

    if (isReachEdge) {
      const nextFlag = {};
      nextFlagRef.current = nextFlag;
      yield doLoadMore(false);
      setFlag(nextFlag);
    }
  }), {
    wait: 100,
    leading: true,
    trailing: true
  });

  _react().default.useEffect(() => {
    check();
  });

  _react().default.useEffect(() => {
    const element = elementRef.current;
    if (!element) return;
    if (!scrollParent) return;

    function onScroll() {
      check();
    }

    scrollParent.addEventListener('scroll', onScroll);
    return () => {
      scrollParent.removeEventListener('scroll', onScroll);
    };
  }, [scrollParent]);

  function retry() {
    return (0, _tslib().__awaiter)(this, void 0, void 0, function* () {
      setFailed(false);
      yield doLoadMore(true);
      setFlag(nextFlagRef.current);
    });
  }

  const renderDone = () => {
    if (props.finishedText) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('finished'))
      }, {
        children: props.finishedText
      }));
    }

    return null;
  };

  const renderFailed = () => {
    if (props.errorText) {
      if (typeof props.errorText === 'function') return props.errorText(retry);
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        onClick: retry,
        className: (0, _clsx().default)(bem('error'))
      }, {
        children: props.errorText
      }));
    }

    return null;
  };

  const renderLoading = () => {
    return (0, _jsxRuntime().jsx)(_loading.default, Object.assign({
      className: (0, _clsx().default)(bem('loading')),
      size: 16
    }, {
      children: props.loadingText || locale.loading
    }));
  };

  const renderPlaceholder = () => {
    if (props.finished) return renderDone();
    if (failed) return renderFailed();
    return renderLoading();
  };

  _react().default.useImperativeHandle(ref, () => ({
    check
  }));

  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem()),
    style: props.style
  }, {
    children: [props.children, (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(props.className, bem('loadmore')),
      ref: elementRef
    }, {
      children: renderPlaceholder()
    }))]
  }));
});

LoadMore.defaultProps = {
  threshold: 300
};
var _default = LoadMore;
exports.default = _default;
