// react-vant/lib/utils/format/unit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addUnit = addUnit;
exports.getSizeStyle = getSizeStyle;
exports.getZIndexStyle = getZIndexStyle;
exports.kebabCase = kebabCase;
exports.unitToPx = unitToPx;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

var _ = require("..");

var _number = require("../validate/number");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function addUnit(value) {
  if (!(0, _.isDef)(value)) {
    return undefined;
  } // eslint-disable-next-line no-param-reassign


  value = String(value);
  return (0, _number.isNumeric)(value) ? `${value}px` : value;
}

function getSizeStyle(originSize) {
  if ((0, _.isDef)(originSize)) {
    const size = addUnit(originSize);
    return {
      width: size,
      height: size
    };
  }

  return {};
}

function getZIndexStyle(zIndex) {
  const style = {};

  if (zIndex !== undefined) {
    style.zIndex = +zIndex;
  }

  return style;
} // cache


let rootFontSize;

function getRootFontSize() {
  if (!rootFontSize) {
    const doc = document.documentElement;
    const fontSize = doc.style.fontSize || window.getComputedStyle(doc).fontSize;
    rootFontSize = parseFloat(fontSize);
  }

  return rootFontSize;
}

function convertRem(value) {
  value = value.replace(/rem/g, '');
  return +value * getRootFontSize();
}

function convertVw(value) {
  value = value.replace(/vw/g, '');
  return +value * window.innerWidth / 100;
}

function convertVh(value) {
  value = value.replace(/vh/g, '');
  return +value * window.innerHeight / 100;
}

function unitToPx(value) {
  if (typeof value === 'number') {
    return value;
  }

  if (_.inBrowser) {
    if (value.indexOf('rem') !== -1) {
      return convertRem(value);
    }

    if (value.indexOf('vw') !== -1) {
      return convertVw(value);
    }

    if (value.indexOf('vh') !== -1) {
      return convertVh(value);
    }
  }

  return parseFloat(value);
}

function kebabCase(str) {
  return str.replace(/([A-Z])/g, '-$1').toLowerCase().replace(/^-/, '');
}
