// react-vant/es/list/LoadMore.js
import { __awaiter } from "tslib";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import clsx from 'clsx';
import { createNamespace } from '../utils';
import useLockFn from '../hooks/use-lock-fn';
import { getScrollParent } from '../hooks/use-scroll-parent';
import Loading from '../loading';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import useThrottleFn from '../hooks/use-throttle-fn';
const [bem] = createNamespace('list');

function isWindow(element) {
  return element === window;
}

const LoadMore = React.forwardRef((props, ref) => {
  const {
    locale
  } = React.useContext(ConfigProviderContext);
  const [failed, setFailed] = React.useState(false);
  const doLoadMore = useLockFn(isRetry => __awaiter(void 0, void 0, void 0, function* () {
    try {
      yield props.onLoad(isRetry);
    } catch (e) {
      setFailed(true);
      throw e;
    }
  }));
  const elementRef = React.useRef(null); // Prevent duplicated trigger of `check` function

  const [flag, setFlag] = React.useState({});
  const nextFlagRef = React.useRef(flag);
  const [scrollParent, setScrollParent] = React.useState();
  const {
    run: check
  } = useThrottleFn(() => __awaiter(void 0, void 0, void 0, function* () {
    if (nextFlagRef.current !== flag) return;
    if (props.finished) return;
    const element = elementRef.current;
    if (!element) return;
    if (!element.offsetParent) return;
    const parent = getScrollParent(element);
    setScrollParent(parent);
    if (!parent) return;
    const rect = element.getBoundingClientRect();
    const elementTop = rect.top;
    const current = isWindow(parent) ? window.innerHeight : parent.getBoundingClientRect().bottom;
    const isReachEdge = current >= elementTop - props.threshold;

    if (isReachEdge) {
      const nextFlag = {};
      nextFlagRef.current = nextFlag;
      yield doLoadMore(false);
      setFlag(nextFlag);
    }
  }), {
    wait: 100,
    leading: true,
    trailing: true
  });
  React.useEffect(() => {
    check();
  });
  React.useEffect(() => {
    const element = elementRef.current;
    if (!element) return;
    if (!scrollParent) return;

    function onScroll() {
      check();
    }

    scrollParent.addEventListener('scroll', onScroll);
    return () => {
      scrollParent.removeEventListener('scroll', onScroll);
    };
  }, [scrollParent]);

  function retry() {
    return __awaiter(this, void 0, void 0, function* () {
      setFailed(false);
      yield doLoadMore(true);
      setFlag(nextFlagRef.current);
    });
  }

  const renderDone = () => {
    if (props.finishedText) {
      return _jsx("div", Object.assign({
        className: clsx(bem('finished'))
      }, {
        children: props.finishedText
      }));
    }

    return null;
  };

  const renderFailed = () => {
    if (props.errorText) {
      if (typeof props.errorText === 'function') return props.errorText(retry);
      return _jsx("div", Object.assign({
        onClick: retry,
        className: clsx(bem('error'))
      }, {
        children: props.errorText
      }));
    }

    return null;
  };

  const renderLoading = () => {
    return _jsx(Loading, Object.assign({
      className: clsx(bem('loading')),
      size: 16
    }, {
      children: props.loadingText || locale.loading
    }));
  };

  const renderPlaceholder = () => {
    if (props.finished) return renderDone();
    if (failed) return renderFailed();
    return renderLoading();
  };

  React.useImperativeHandle(ref, () => ({
    check
  }));
  return _jsxs("div", Object.assign({
    className: clsx(props.className, bem()),
    style: props.style
  }, {
    children: [props.children, _jsx("div", Object.assign({
      className: clsx(props.className, bem('loadmore')),
      ref: elementRef
    }, {
      children: renderPlaceholder()
    }))]
  }));
});
LoadMore.defaultProps = {
  threshold: 300
};
export default LoadMore;
