// react-vant/lib/field/Field.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FIELD_KEY = void 0;

function _tslib() {
  const data = require("tslib");

  _tslib = function () {
    return data;
  };

  return data;
}

function _jsxRuntime() {
  const data = require("react/jsx-runtime");

  _jsxRuntime = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireWildcard(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _icons() {
  const data = require("@react-vant/icons");

  _icons = function () {
    return data;
  };

  return data;
}

function _clsx() {
  const data = _interopRequireDefault(require("clsx"));

  _clsx = function () {
    return data;
  };

  return data;
}

var _cell = _interopRequireDefault(require("../cell"));

var _dialog = _interopRequireDefault(require("../dialog"));

var _input = _interopRequireDefault(require("../input"));

var _utils = require("../utils");

var _constant = require("../utils/constant");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const [bem] = (0, _utils.createNamespace)('field');
const Field = (0, _react().forwardRef)((props, ref) => {
  var _a;

  const inputRef = (0, _react().useRef)(null);
  const textareaRef = (0, _react().useRef)(null);
  const elementRef = props.type === 'textarea' ? textareaRef : inputRef;

  const focus = () => {
    if (elementRef.current) {
      elementRef.current.focus();
    }
  };

  const blur = () => {
    if (elementRef.current) {
      elementRef.current.blur();
    }
  };

  const clear = () => {
    if (elementRef.current) {
      elementRef.current.clear();
    }
  };

  (0, _react().useImperativeHandle)(ref, () => ({
    focus,
    blur,
    clear,

    get nativeElement() {
      return elementRef.current.nativeElement;
    }

  }));

  const getProp = key => {
    if ((0, _utils.isDef)(props[key])) {
      return props[key];
    }

    return null;
  };

  const labelStyle = () => {
    const labelW = getProp('labelWidth');

    if (labelW) {
      return {
        width: (0, _utils.addUnit)(labelW)
      };
    }

    return {};
  };

  const formatValue = (inputValue, trigger = 'onChange') => {
    const {
      formatTrigger,
      formatter
    } = props;

    if (formatter && trigger === formatTrigger) {
      return formatter(inputValue);
    }

    return inputValue;
  };

  const onChange = val => {
    var _a;

    (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, formatValue(val));
  };

  const renderInput = () => {
    const {
      value,
      defaultValue,
      align,
      type,
      placeholder,
      name,
      maxLength,
      disabled,
      readOnly,
      clearable,
      clearIcon,
      clearTrigger,
      autoFocus,
      onClear,
      onBlur,
      onFocus,
      onKeyPress,
      onOverlimit
    } = props;

    if (props.children) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('children'))
      }, {
        children: props.children
      }));
    }

    const commonProps = {
      value,
      onChange,
      placeholder,
      name,
      defaultValue,
      disabled,
      clearable,
      clearIcon,
      clearTrigger,
      onClear,
      onBlur,
      onFocus,
      onKeyPress,
      onOverlimit,
      autoFocus,
      readOnly,
      maxLength,
      onClick: props.onClickInput
    };

    if (type === 'textarea') {
      return (0, _jsxRuntime().jsx)(_input.default.TextArea, Object.assign({
        ref: textareaRef,
        autoSize: props.autoSize,
        showWordLimit: props.showWordLimit,
        rows: props.rows
      }, commonProps));
    }

    return (0, _jsxRuntime().jsx)(_input.default, Object.assign({
      ref: inputRef,
      type: type,
      align: align
    }, commonProps));
  };

  const renderLeftIcon = () => {
    const {
      leftIcon,
      onClickLeftIcon
    } = props;
    if (!leftIcon) return null;
    return (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('left-icon')),
      onClick: onClickLeftIcon
    }, {
      children: leftIcon
    }));
  };

  const renderRightIcon = () => {
    const {
      rightIcon,
      onClickRightIcon
    } = props;
    if (!rightIcon) return null;
    return (0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('right-icon')),
      onClick: onClickRightIcon
    }, {
      children: rightIcon
    }));
  };

  const renderMessage = () => {
    const message = props.errorMessage;

    if (message) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('error-message'))
      }, {
        children: message
      }));
    }

    return null;
  };

  const renderIntro = () => {
    if (props.intro) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('intro'))
      }, {
        children: props.intro
      }));
    }

    return null;
  };

  const renderTooltip = () => {
    const {
      tooltip
    } = props;

    if (tooltip) {
      let icon = (0, _jsxRuntime().jsx)(_icons().QuestionO, {});
      let dialogProps = {
        message: tooltip
      };

      if (!(_react().default.isValidElement(tooltip) || typeof tooltip === 'string')) {
        const _a = tooltip,
              {
          icon: customIcon
        } = _a,
              customDialogProps = (0, _tslib().__rest)(_a, ["icon"]);
        icon = customIcon || icon;
        dialogProps = customDialogProps;
      }

      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('tooltip')),
        onClick: () => _dialog.default.show(dialogProps)
      }, {
        children: icon
      }));
    }

    return null;
  };

  const renderLabel = () => {
    const {
      label,
      colon
    } = props;

    if (label) {
      return (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
        children: [label, !!colon && ':', renderTooltip()]
      });
    }

    return null;
  };

  const {
    size,
    center,
    border,
    isLink,
    required,
    clickable,
    labelAlign,
    className,
    labelClass,
    valueClass,
    controlAlign,
    arrowDirection,
    disabled,
    titleStyle,
    error
  } = props;
  const suffix = (_a = props.suffix) !== null && _a !== void 0 ? _a : props.button;
  return (0, _jsxRuntime().jsxs)(_cell.default, Object.assign({
    title: renderLabel(),
    size: size,
    icon: renderLeftIcon(),
    center: center,
    border: border,
    isLink: isLink,
    required: required,
    clickable: clickable,
    extra: props.extra,
    titleStyle: Object.assign(Object.assign({}, labelStyle()), titleStyle),
    valueClass: (0, _clsx().default)(bem('value', [controlAlign]), valueClass),
    titleClass: (0, _clsx().default)(bem('label', labelAlign), labelClass),
    arrowDirection: arrowDirection,
    onClick: props.onClick,
    style: props.style,
    className: (0, _clsx().default)(bem({
      error,
      disabled,
      [`label-${labelAlign}`]: labelAlign
    }), className)
  }, {
    children: [(0, _jsxRuntime().jsxs)("div", Object.assign({
      className: (0, _clsx().default)(bem('body'))
    }, {
      children: [props.prefix && (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('prefix'))
      }, {
        children: props.prefix
      })), (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('control-wrapper'))
      }, {
        children: renderInput()
      })), renderRightIcon(), suffix && (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('suffix'))
      }, {
        children: suffix
      }))]
    })), renderMessage(), renderIntro()]
  }));
});
Field.defaultProps = {
  clearIcon: (0, _jsxRuntime().jsx)(_icons().Clear, {}),
  clearTrigger: 'focus',
  formatTrigger: 'onChange',
  defaultValue: ''
};
const FIELD_KEY = Symbol('field');
exports.FIELD_KEY = FIELD_KEY;
const FieldNamespace = Object.assign(Field, {
  [_constant.COMPONENT_TYPE_KEY]: FIELD_KEY
});
var _default = FieldNamespace;
exports.default = _default;
